<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../.resources/report.css" type="text/css"/><link rel="shortcut icon" href="../.resources/report.gif" type="image/gif"/><title>ConsedAssemblyTransformerBuilder.java</title><link rel="stylesheet" href="../.resources/prettify.css" type="text/css"/><script type="text/javascript" src="../.resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../.sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">Jillion</a> &gt; <a href="index.source.html" class="el_package">org.jcvi.jillion.assembly.consed</a> &gt; <span class="el_source">ConsedAssemblyTransformerBuilder.java</span></div><h1>ConsedAssemblyTransformerBuilder.java</h1><pre class="source lang-java linenums">/*******************************************************************************
 * Jillion development code
 * 
 * This code may be freely distributed and modified under the
 * terms of the GNU Lesser General Public License.  This should
 * be distributed with the code.  If you do not have a copy,
 *  see:
 * 
 *          http://www.gnu.org/copyleft/lesser.html
 * 
 * 
 * Copyright for this code is held jointly by the individual authors.  These should be listed in the @author doc comments.
 * 
 * Information about Jillion can be found on its homepage
 * 
 *         http://jillion.sourceforge.net
 * 
 * Contributors:
 *     Danny Katzel - initial API and implementation
 ******************************************************************************/
package org.jcvi.jillion.assembly.consed;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.URI;
import java.time.Clock;
import java.util.Arrays;
import java.util.Date;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Objects;

import org.jcvi.jillion.assembly.AssemblyTransformer;
import org.jcvi.jillion.assembly.ReadInfo;
import org.jcvi.jillion.assembly.consed.ace.AceContig;
import org.jcvi.jillion.assembly.consed.ace.AceContigBuilder;
import org.jcvi.jillion.assembly.consed.ace.AceFileWriter;
import org.jcvi.jillion.assembly.consed.ace.AceFileWriterBuilder;
import org.jcvi.jillion.assembly.consed.ace.PhdInfo;
import org.jcvi.jillion.assembly.consed.ace.WholeAssemblyAceTag;
import org.jcvi.jillion.assembly.consed.phd.PhdBuilder;
import org.jcvi.jillion.assembly.consed.phd.PhdDataStore;
import org.jcvi.jillion.assembly.consed.phd.PhdUtil;
import org.jcvi.jillion.core.Direction;
import org.jcvi.jillion.core.Range;
import org.jcvi.jillion.core.io.FileUtil;
import org.jcvi.jillion.core.io.IOUtil;
import org.jcvi.jillion.core.pos.PositionSequence;
import org.jcvi.jillion.core.qual.PhredQuality;
import org.jcvi.jillion.core.qual.QualitySequence;
import org.jcvi.jillion.core.qual.QualitySequenceBuilder;
import org.jcvi.jillion.core.residue.nt.NucleotideSequence;
/**
 * {@code ConsedAssemblyTransformerBuilder}
 * will build an {@link AssemblyTransformer}
 * instance that will convert an assembly
 * into a consed package (ace file, phd file(s) etc).
 * 
 * @author dkatzel
 *
 */
public class ConsedAssemblyTransformerBuilder {
	/**
	 * The default quality value to use if qualities not provided: 26.
	 */
<span class="nc" id="L74">	public static final PhredQuality DEFAULT_QUALITY_VALUE = PhredQuality.valueOf(26);</span>
	private final File rootDir;
	private final String filePrefix;
	
<span class="nc" id="L78">	private boolean createPhdBall = true;</span>
	
<span class="nc" id="L80">	private File chromatInputDir = null;</span>
<span class="nc" id="L81">	private byte defaultQualityValue = DEFAULT_QUALITY_VALUE.getQualityScore();</span>
	
	
<span class="nc" id="L84">	private Clock clock = Clock.systemDefaultZone();</span>
	
<span class="nc" id="L86">	private AceAssemblyTransformerPostProcessor postProcessor =null;</span>
	/**
	 * Create a new ConsedAssemblyTransformerBuilder instance that will
	 * write out a consed package using the given parameters.
	 * @param consedRootDir the root directory of the consed package to write.
	 * Underneath this directory, this transformer will create and populate &quot;edit_dir&quot; and &quot;phd_dir&quot; or
	 * &quot;phdball_dir&quot; directories. if the consedRootDir does not yet exist,
	 * then the directory will be created. Can not be null.
	 * @param filePrefix the file prefix to use for the ace and phd files to write.
	 * These files will named $filePrefix.ace.1 and $filePrefix.phd.ball.1 etc.
	 * @throws NullPointerException if any parameters are null.  Can not be null.
	 */
<span class="nc" id="L98">	public ConsedAssemblyTransformerBuilder(File consedRootDir, String filePrefix){</span>
		
<span class="nc bnc" id="L100" title="All 2 branches missed.">		if(consedRootDir ==null){</span>
<span class="nc" id="L101">			throw new NullPointerException(&quot;root dir can not be null&quot;);</span>
		}
<span class="nc bnc" id="L103" title="All 2 branches missed.">		if(filePrefix ==null){</span>
<span class="nc" id="L104">			throw new NullPointerException(&quot;file prefix can not be null&quot;);</span>
		}
<span class="nc bnc" id="L106" title="All 2 branches missed.">		if(filePrefix.isEmpty()){</span>
<span class="nc" id="L107">			throw new IllegalArgumentException(&quot;file prefix can not be empty&quot;);</span>
		}
<span class="nc" id="L109">		this.rootDir = consedRootDir;</span>
<span class="nc" id="L110">		this.filePrefix = filePrefix;</span>
		
<span class="nc" id="L112">	}</span>
	/**
	 * Set the {@link Clock} to use for the timestamps
	 * for various assembly metadata.
	 * @param clock a {@link Clock} instance; can not be null.
	 * @return this
	 * 
	 * @throws NullPointerException if clock is null.
	 */
	public ConsedAssemblyTransformerBuilder setClock(Clock clock){
<span class="nc" id="L122">	    Objects.requireNonNull(clock);</span>
<span class="nc" id="L123">	    this.clock=clock;</span>
<span class="nc" id="L124">	    return this;</span>
	}
	/**
	 * Should individual phd files in the &quot;phd_dir&quot; directory be created instead of using one giant
	 * phd.ball file.  Creating individual files
	 * is not recommended for next-gen assemblies since there will potentially be millions of 
	 * reads and therefore millions of files in a single directory.  However, this option is provided
	 * so that consed packages can be used on legacy consed installations or other ace and phd
	 * parser libraries and tools that don't support phd.ball.
	 * If this method is not called, then the default
	 * is to use a phd.ball file (similar to {@link #createIndividualPhdFiles(boolean) createIndividualPhdFiles(false)}
	 * @param createIndividualPhdFiles {@code true} if should create individual phd files in the phd_dir; {@code false}
	 * if should use phd.ball in phdball_dir instead (default is {@code false}.
	 * @return this.
	 */
	public ConsedAssemblyTransformerBuilder createIndividualPhdFiles(boolean createIndividualPhdFiles){
<span class="nc bnc" id="L140" title="All 2 branches missed.">		this.createPhdBall = !createIndividualPhdFiles;</span>
<span class="nc" id="L141">		return this;</span>
	}
	/**
	 * Provide a directory which contains sanger chromatograms
	 * to be included in this consed package.  If the directory is not null
	 * and contains sanger trace files, then those traces will be copied and coverted into 
	 * scf format under $consed_rootDir/chromat_dir so consed can see them and display the
	 * raw chromatogram channel wave forms. 
	 * @param chromatDir the chormatDir to get the traces from;
	 * if {@code null} then there are no chroamtograms (defaults to {@code null}
	 * @return this
	 */
	public ConsedAssemblyTransformerBuilder inputChromatogramDir(File chromatDir){
<span class="nc" id="L154">		this.chromatInputDir = chromatDir;</span>
<span class="nc" id="L155">		return this;</span>
	}
	/**
	 * Set the {@link AceAssemblyTransformerPostProcessor} to use
	 * to modify the Ace contigs after the transformation service
	 * has provided all the alignment information.
	 * @param postProcessor the {@link AceAssemblyTransformerPostProcessor}
	 * to use; or {@code null} if there is no post processor (defaults to {@code null} ).
	 * @return this
	 */
	public ConsedAssemblyTransformerBuilder postProcessor(AceAssemblyTransformerPostProcessor postProcessor){
<span class="nc" id="L166">		this.postProcessor = postProcessor;</span>
<span class="nc" id="L167">		return this;</span>
	}
	/**
	 * Set the default quality value to use if an aligned read
	 * does not have associated quality data.  If not called,
	 * then the default is {@link #DEFAULT_QUALITY_VALUE}
	 * @param qualityValue the quality value to use;
	 * must be between 0 and {@link Byte#MAX_VALUE}.
	 * @return this.
	 * @throws IllegalArgumentException if if qualityScore &amp;lt; 0 or &amp;gt; {@link Byte#MAX_VALUE}.
	 */
	public ConsedAssemblyTransformerBuilder setDefaultQualityValue(int qualityValue){
		//Phredquality object does all validation for us
<span class="nc" id="L180">		PhredQuality qual = PhredQuality.valueOf(qualityValue);</span>
<span class="nc" id="L181">		this.defaultQualityValue = qual.getQualityScore();</span>
<span class="nc" id="L182">		return this;</span>
	}
	/**
	 * Create a new {@link AssemblyTransformer} instance
	 * using the provided configuration.
	 * @return a new {@link AssemblyTransformer}
	 * will never be null.
	 * @throws IOException if there is a problem creating the consed root dir
	 * or any of the underlying files (which may initially be empty until
	 * the {@link AssemblyTransformer} methods have been called.
	 */
	public AssemblyTransformer build() throws IOException{
<span class="nc" id="L194">		return new ConsedAssemblyTransformer(this);</span>
	}
	
	private static final class ConsedAssemblyTransformer implements AssemblyTransformer{
		private final File rootDir, editDir, chromatInputDir, chromatDir;
		private final String filePrefix;
		
		private final AceAssemblyTransformerPostProcessor postProcessor;
		
		private final PhdConsedTransformerHelper phdHelper;
		private final byte defaultQualityValue;
		
<span class="nc" id="L206">		private final Map&lt;String, AceContigBuilder&gt; builderMap = new LinkedHashMap&lt;String, AceContigBuilder&gt;();</span>
		
<span class="nc" id="L208">		private final Map&lt;URI,Date&gt; uriDates = new HashMap&lt;URI, Date&gt;();</span>
<span class="nc" id="L209">		private final Map&lt;URI,File&gt; uri2File = new HashMap&lt;URI, File&gt;();</span>
		
<span class="nc" id="L211">		private final Map&lt;URI,Map&lt;String,String&gt;&gt; comments = new HashMap&lt;URI, Map&lt;String,String&gt;&gt;();</span>
		
		
<span class="nc" id="L214">		public ConsedAssemblyTransformer(ConsedAssemblyTransformerBuilder builder) throws IOException{</span>
			
			
<span class="nc" id="L217">			this.rootDir = builder.rootDir;</span>
<span class="nc" id="L218">			this.filePrefix = builder.filePrefix;</span>
<span class="nc" id="L219">			this.chromatInputDir = builder.chromatInputDir;</span>
<span class="nc bnc" id="L220" title="All 2 branches missed.">			this.postProcessor = builder.postProcessor ==null? NullAceAssemblyTransformerPostProcessor.INSTANCE : builder.postProcessor;</span>
		
<span class="nc" id="L222">			IOUtil.mkdirs(rootDir);</span>
<span class="nc" id="L223">			this.editDir = new File(rootDir, &quot;edit_dir&quot;);</span>
<span class="nc bnc" id="L224" title="All 2 branches missed.">			if(chromatInputDir ==null){</span>
<span class="nc" id="L225">				this.chromatDir = null;				</span>
			}else{
				//has chromats
<span class="nc" id="L228">				this.chromatDir = new File(rootDir, &quot;chromat_dir&quot;);</span>
<span class="nc" id="L229">				IOUtil.mkdirs(chromatDir);</span>
				//TODO copy chromatograms? to chromat_dir?
			}
<span class="nc bnc" id="L232" title="All 2 branches missed.">			if(builder.createPhdBall){</span>
<span class="nc" id="L233">				phdHelper = new PhdBallConsedTransformerHelper(rootDir);</span>
			}else{
<span class="nc" id="L235">				phdHelper = new IndividualPhdConsedTransformerHelper(rootDir);</span>
			}
			
<span class="nc" id="L238">			this.defaultQualityValue = builder.defaultQualityValue;</span>
			
			//add default
<span class="nc" id="L241">			Date currentDate = new Date(builder.clock.millis());</span>
<span class="nc" id="L242">			uriDates.put(null, currentDate);</span>
<span class="nc" id="L243">			comments.put(null, computeRequiredCommentsFor(null, currentDate));</span>
			
<span class="nc" id="L245">		}</span>

		@Override
		public void referenceOrConsensus(String id,
				NucleotideSequence gappedReference) {
<span class="nc bnc" id="L250" title="All 2 branches missed.">			if(builderMap.containsKey(id)){</span>
<span class="nc" id="L251">				throw new IllegalStateException(&quot;reference with id &quot; + id + &quot; already exists&quot;);</span>
			}
<span class="nc" id="L253">			builderMap.put(id, new AceContigBuilder(id, gappedReference));</span>
			
<span class="nc" id="L255">		}</span>

		@Override
		public void endAssembly() {
			final PhdDataStore phdDataStore;
			try {
<span class="nc" id="L261">				phdDataStore =phdHelper.createDataStore();</span>
<span class="nc" id="L262">			} catch (IOException e) {</span>
<span class="nc" id="L263">				throw new IllegalStateException(&quot;error reading phd data&quot;, e);</span>
<span class="nc" id="L264">			}</span>
<span class="nc" id="L265">			Map&lt;String, AceContigBuilder&gt; updatedBuilders = postProcessor.postProcess(builderMap, phdDataStore);</span>
			
<span class="nc" id="L267">			File outputAceFile = new File(editDir, String.format(&quot;%s.ace.1&quot;,filePrefix ));</span>
			final AceFileWriter aceWriter;
			try {
<span class="nc" id="L270">				aceWriter = new AceFileWriterBuilder(outputAceFile, phdDataStore)</span>
<span class="nc" id="L271">											.build();</span>
<span class="nc" id="L272">			} catch (IOException e) {</span>
<span class="nc" id="L273">				throw new IllegalStateException(&quot;error creating ace writer&quot;, e);</span>
<span class="nc" id="L274">			}</span>
			try{
<span class="nc bnc" id="L276" title="All 2 branches missed.">				for(AceContigBuilder builder : updatedBuilders.values()){</span>
					try {
<span class="nc bnc" id="L278" title="All 2 branches missed.">						for(Entry&lt;Range,AceContig&gt; entry : ConsedUtil.split0xContig(builder,true).entrySet()){</span>
<span class="nc" id="L279">		                    AceContig splitContig = entry.getValue();</span>
<span class="nc" id="L280">		                    postProcessor.postProcess(splitContig);</span>
<span class="nc" id="L281">		                    aceWriter.write(splitContig);</span>
<span class="nc" id="L282">						}</span>
					
						
<span class="nc" id="L285">					} catch (IOException e) {</span>
<span class="nc" id="L286">						throw new IllegalStateException(&quot;error writing assembly &quot; + builder.getContigId(), e);</span>
<span class="nc" id="L287">					}</span>
<span class="nc" id="L288">				}</span>
				
<span class="nc" id="L290">				WholeAssemblyAceTag phdBallTag = phdHelper.createPhdBallWholeAssemblyTag();</span>
<span class="nc bnc" id="L291" title="All 2 branches missed.">				if(phdBallTag !=null){</span>
<span class="nc" id="L292">					aceWriter.write(phdBallTag);</span>
				}	
<span class="nc" id="L294">			} catch (IOException e) {</span>
<span class="nc" id="L295">				throw new IllegalStateException(&quot;error creating ace tag&quot;, e);</span>
			}finally{
<span class="nc" id="L297">				IOUtil.closeAndIgnoreErrors(aceWriter);</span>
<span class="nc" id="L298">			}</span>
			
<span class="nc" id="L300">		}</span>

		@Override
		public void notAligned(String id,
				NucleotideSequence nucleotideSequence,
				QualitySequence qualitySequence, PositionSequence positions,
				URI uri) {
			//don't care
			
<span class="nc" id="L309">		}</span>

		@Override
		public void aligned(String readId, NucleotideSequence nucleotideSequence,
				QualitySequence qualitySequence, PositionSequence positions,
				URI sourceFileUri, String referenceId, long gappedStartOffset,
				Direction direction,
				NucleotideSequence gappedSequence,
				ReadInfo readInfo) {
			
			//TODO handle symlink to uri?
<span class="nc bnc" id="L320" title="All 2 branches missed.">			if(!builderMap.containsKey(referenceId)){</span>
<span class="nc" id="L321">				throw new IllegalStateException(&quot;unknown reference id &quot; + referenceId);</span>
			}
<span class="nc bnc" id="L323" title="All 2 branches missed.">			if(gappedStartOffset &gt; Integer.MAX_VALUE){</span>
				//TODO Violation of LSP ?
<span class="nc" id="L325">				throw new IllegalArgumentException(&quot;gapped start offset is &gt; int max&quot;+ readId);</span>
			}
			final QualitySequence qualities;
<span class="nc bnc" id="L328" title="All 2 branches missed.">			if(qualitySequence ==null){</span>
<span class="nc" id="L329">				qualities = createDefaultQualitySequenceFor(nucleotideSequence);</span>
			}else{
<span class="nc" id="L331">				qualities = qualitySequence;</span>
			}
			final Date phdDate;
			Map&lt;String, String&gt; requiredComments;
			
<span class="nc bnc" id="L336" title="All 2 branches missed.">			if(uriDates.containsKey(sourceFileUri)){</span>
<span class="nc" id="L337">				phdDate = uriDates.get(sourceFileUri);</span>
<span class="nc" id="L338">				requiredComments = comments.get(sourceFileUri);</span>
			}else{
<span class="nc" id="L340">				File file = new File(sourceFileUri);</span>
<span class="nc" id="L341">				phdDate = new Date(file.lastModified());</span>
<span class="nc" id="L342">				uriDates.put(sourceFileUri, phdDate);</span>
<span class="nc" id="L343">				uri2File.put(sourceFileUri, file);</span>
				
				
				//first time we've seen this file
<span class="nc bnc" id="L347" title="All 2 branches missed.">				if(file.getName().endsWith(&quot;.scf&quot;)){</span>
<span class="nc" id="L348">					InputStream in=null;</span>
<span class="nc" id="L349">					OutputStream out = null;</span>
					try{
<span class="nc" id="L351">						in = new BufferedInputStream(new FileInputStream(file));</span>
<span class="nc" id="L352">						File newFile = new File(chromatDir, FileUtil.getBaseName(file));</span>
<span class="nc" id="L353">						out = new BufferedOutputStream(new FileOutputStream(newFile));</span>
<span class="nc" id="L354">						IOUtil.copy(in, out);</span>
<span class="nc" id="L355">					} catch (IOException e) {</span>
<span class="nc" id="L356">						throw new IllegalStateException(&quot;error copying chromatogram file to chromat_dir&quot;, e);</span>
					}finally{
<span class="nc" id="L358">						IOUtil.closeAndIgnoreErrors(in, out);</span>
<span class="nc" id="L359">					}</span>
				}
<span class="nc" id="L361">				requiredComments = computeRequiredCommentsFor(uri2File.get(sourceFileUri), phdDate);</span>
<span class="nc" id="L362">				comments.put(sourceFileUri, requiredComments);</span>
			}
<span class="nc" id="L364">			PhdBuilder phdBuilder = new PhdBuilder(readId, nucleotideSequence, qualities)</span>
<span class="nc" id="L365">										.comments(requiredComments);</span>
			
<span class="nc bnc" id="L367" title="All 2 branches missed.">			if(positions ==null){</span>
<span class="nc" id="L368">				phdBuilder.fakePeaks();			</span>
			}else{				
<span class="nc" id="L370">				phdBuilder.peaks(positions);</span>
			}
			final PhdInfo phdInfo;
			try {
<span class="nc" id="L374">				phdInfo = phdHelper.writePhd(phdBuilder.build(), phdDate);</span>
<span class="nc" id="L375">			} catch (IOException e) {</span>
<span class="nc" id="L376">				throw new IllegalStateException(&quot;error writing phd record for &quot;+ readId, e);</span>
<span class="nc" id="L377">			}</span>
			
<span class="nc" id="L379">			builderMap.get(referenceId)</span>
<span class="nc" id="L380">							.addRead(readId, gappedSequence, (int)gappedStartOffset, direction, readInfo.getValidRange(), </span>
<span class="nc" id="L381">									phdInfo, readInfo.getUngappedFullLength());</span>
			
<span class="nc" id="L383">		}</span>

		private Map&lt;String,String&gt; computeRequiredCommentsFor(File file, Date phdDate ){
<span class="nc bnc" id="L386" title="All 2 branches missed.">			if(file ==null){</span>
<span class="nc" id="L387">				return PhdUtil.createPhdTimeStampCommentFor(phdDate);</span>
			}
			try{
<span class="nc bnc" id="L390" title="All 2 branches missed.">				if(isChromatogramFile(file)){</span>
<span class="nc" id="L391">					return PhdUtil.createPhdTimeStampAndChromatFileCommentsFor(phdDate, file.getName());</span>
				}
<span class="nc" id="L393">				return PhdUtil.createPhdTimeStampCommentFor(phdDate);</span>
<span class="nc" id="L394">			}catch(Exception e){</span>
<span class="nc" id="L395">				return PhdUtil.createPhdTimeStampCommentFor(phdDate);</span>
			}

			
			
		}
		
		private boolean isChromatogramFile(File f){
<span class="nc bnc" id="L403" title="All 4 branches missed.">			return !f.getName().endsWith(&quot;.fastq&quot;) &amp;&amp; !f.getName().endsWith(&quot;.sff&quot;);</span>
			
		}
		
		 private QualitySequence createDefaultQualitySequenceFor(NucleotideSequence rawSequence) {
<span class="nc" id="L408">		        int numberOfQualities =(int) rawSequence.getUngappedLength();</span>
<span class="nc" id="L409">				byte[] qualities = new byte[numberOfQualities];</span>
<span class="nc" id="L410">				Arrays.fill(qualities, defaultQualityValue);</span>
<span class="nc" id="L411">		        return new QualitySequenceBuilder(qualities).build();</span>
		    }

		@Override
		public void assemblyCommand(String name, String version,
				String parameters) {
			//no-op
			
<span class="nc" id="L419">		}</span>

		@Override
		public void referenceFile(URI uri) {
			//no-op
			
<span class="nc" id="L425">		}</span>

		@Override
		public void readFile(URI uri) {
			//TODO handle symlink to uri?
			
<span class="nc" id="L431">		}</span>
		
		
	}
	
<span class="nc" id="L436">	private enum NullAceAssemblyTransformerPostProcessor implements AceAssemblyTransformerPostProcessor{</span>
<span class="nc" id="L437">		INSTANCE;</span>

		@Override
		public Map&lt;String, AceContigBuilder&gt; postProcess(
				Map&lt;String, AceContigBuilder&gt; builderMap, PhdDataStore phdDataStore) {
<span class="nc" id="L442">			return builderMap;</span>
		}

		@Override
		public void postProcess(AceContig contig) {
			//no-op
			
<span class="nc" id="L449">		}</span>
		
		
		
	}
	
	
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.5.201505241946</span></div></body></html>