<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../.resources/report.css" type="text/css"/><link rel="shortcut icon" href="../.resources/report.gif" type="image/gif"/><title>AbstractReadCasVisitor.java</title><link rel="stylesheet" href="../.resources/prettify.css" type="text/css"/><script type="text/javascript" src="../.resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../.sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">Jillion</a> &gt; <a href="index.source.html" class="el_package">org.jcvi.jillion.assembly.clc.cas</a> &gt; <span class="el_source">AbstractReadCasVisitor.java</span></div><h1>AbstractReadCasVisitor.java</h1><pre class="source lang-java linenums">/*******************************************************************************
 * Jillion development code
 * 
 * This code may be freely distributed and modified under the
 * terms of the GNU Lesser General Public License.  This should
 * be distributed with the code.  If you do not have a copy,
 *  see:
 * 
 *          http://www.gnu.org/copyleft/lesser.html
 * 
 * 
 * Copyright for this code is held jointly by the individual authors.  These should be listed in the @author doc comments.
 * 
 * Information about Jillion can be found on its homepage
 * 
 *         http://jillion.sourceforge.net
 * 
 * Contributors:
 *     Danny Katzel - initial API and implementation
 ******************************************************************************/
package org.jcvi.jillion.assembly.clc.cas;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import org.jcvi.jillion.core.datastore.DataStore;
import org.jcvi.jillion.core.datastore.DataStoreException;
import org.jcvi.jillion.core.datastore.DataStoreProviderHint;
import org.jcvi.jillion.core.io.IOUtil;
import org.jcvi.jillion.core.qual.PhredQuality;
import org.jcvi.jillion.core.qual.QualitySequence;
import org.jcvi.jillion.core.qual.QualitySequenceBuilder;
import org.jcvi.jillion.core.residue.nt.NucleotideSequence;
import org.jcvi.jillion.core.util.iter.IteratorUtil;
import org.jcvi.jillion.core.util.iter.StreamingIterator;
import org.jcvi.jillion.fasta.nt.NucleotideFastaDataStore;
import org.jcvi.jillion.fasta.nt.NucleotideFastaFileDataStoreBuilder;
import org.jcvi.jillion.trace.Trace;
import org.jcvi.jillion.trace.TraceDataStore;
import org.jcvi.jillion.trace.fastq.FastqFileDataStore;
import org.jcvi.jillion.trace.fastq.FastqFileDataStoreBuilder;
import org.jcvi.jillion.trace.fastq.FastqQualityCodec;
import org.jcvi.jillion.trace.fastq.FastqRecord;
import org.jcvi.jillion.trace.sff.SffFileDataStoreBuilder;
/**
 * A {@link CasFileVisitor}
 * that handles iterating over the aligned reads in the cas file.
 * Cas files don't actually store any read information, just pointers
 * to where the input read files exist on the file system.
 * This means that all the read alignments in the cas must be matched up 
 * exactly with the input read files.  This can get very complicated
 * so this class handles all of that for you.
 * &lt;p&gt;
 * This abstract class finds all the read files to be parsed
 * and maps them to the corresponding read in the input sequence file(s)
 * so you can get the {@link Trace} for each read.
 * For each read that aligned (matched), the aligned()
 * method is called; for each non-aligned read, the notAligned()
 * is called.
 * 
 * @author dkatzel
 *
 * @since 5.1
 */
public abstract class AbstractReadCasVisitor extends AbstractCasFileVisitor{

	private final File workingDir;
	
<span class="fc" id="L74">	private final List&lt;StreamingIterator&lt;? extends Trace&gt;&gt; iterators = new ArrayList&lt;StreamingIterator&lt;? extends Trace&gt;&gt;();</span>
	
<span class="fc" id="L76">	private FastqQualityCodec qualityCodec=null;</span>
	
<span class="fc" id="L78">	private List&lt;String&gt; referenceIds = new ArrayList&lt;&gt;();</span>
	
	private StreamingIterator&lt;Trace&gt; chainedTraceIterator;
	
	
	/**
	 * Create a new instance.
	 * 
	 * @param workingDir the working directory of the cas file;
	 * may be null (means current directory).
	 * 
	 * @throws NullPointerException if gappedReferenceDataStore is null.
	 */
<span class="fc" id="L91">	public AbstractReadCasVisitor(File workingDir) {</span>
		
<span class="fc" id="L93">		this.workingDir = workingDir;</span>
		
<span class="fc" id="L95">	}</span>

	
	
	public FastqQualityCodec getQualityCodec() {
<span class="nc" id="L100">		return qualityCodec;</span>
	}



	public void setQualityCodec(FastqQualityCodec qualityCodec) {
<span class="nc" id="L106">		this.qualityCodec = qualityCodec;</span>
<span class="nc" id="L107">	}</span>



	public File getWorkingDir() {
<span class="fc" id="L112">		return workingDir;</span>
	}


	@Override
	public void visitReferenceFileInfo(CasFileInfo referenceFileInfo) {
<span class="fc bfc" id="L118" title="All 2 branches covered.">		for(String path: referenceFileInfo.getFileNames()){</span>
			try{
<span class="fc" id="L120">				File fastaFile = CasUtil.getFileFor(workingDir, path);</span>
				
<span class="pc" id="L122">				try(NucleotideFastaDataStore datastore = new NucleotideFastaFileDataStoreBuilder(fastaFile)</span>
<span class="fc" id="L123">															.hint(DataStoreProviderHint.ITERATION_ONLY)</span>
<span class="fc" id="L124">															.build();</span>
<span class="fc" id="L125">					StreamingIterator&lt;String&gt; iter = datastore.idIterator();</span>
						){
<span class="fc bfc" id="L127" title="All 2 branches covered.">					while(iter.hasNext()){</span>
<span class="fc" id="L128">						referenceIds.add(iter.next());</span>
					}
<span class="pc bpc" id="L130" title="12 of 16 branches missed.">				}</span>
<span class="nc" id="L131">			}catch(Exception e){</span>
<span class="nc" id="L132">				throw new IllegalStateException(&quot;error parsing reference data&quot;, e);</span>
<span class="fc" id="L133">			}</span>
<span class="fc" id="L134">		}</span>
<span class="fc" id="L135">	}</span>



	@Override
	public void visitReadFileInfo(CasFileInfo readFileInfo) {
<span class="fc" id="L141">		List&lt;String&gt; fileNames = readFileInfo.getFileNames();</span>
<span class="pc bpc" id="L142" title="1 of 2 branches missed.">		if(fileNames.isEmpty()){</span>
			//nothing to do... 
<span class="nc" id="L144">			return;</span>
		}
<span class="pc bpc" id="L146" title="1 of 2 branches missed.">		if(fileNames.size()==1){</span>
			try {
<span class="fc" id="L148">				File file = CasUtil.getFileFor(workingDir, fileNames.get(0));</span>
				
<span class="fc" id="L150">				iterators.add(createIteratorFor(file));</span>
<span class="nc" id="L151">			} catch (Exception e) {</span>
<span class="nc bnc" id="L152" title="All 2 branches missed.">				for(StreamingIterator&lt;? extends Trace&gt; iter : iterators){</span>
<span class="nc" id="L153">					IOUtil.closeAndIgnoreErrors(iter);</span>
<span class="nc" id="L154">				}</span>
<span class="nc" id="L155">				throw new IllegalStateException(&quot;error getting input read data&quot;, e);</span>
<span class="fc" id="L156">			}      </span>
		}else{
<span class="nc" id="L158">			List&lt;StreamingIterator&lt;? extends Trace&gt;&gt; iteratorLists = new ArrayList&lt;StreamingIterator&lt;? extends Trace&gt;&gt;(fileNames.size());</span>
			try {
<span class="nc bnc" id="L160" title="All 2 branches missed.">				for(String filePath :fileNames){					</span>
<span class="nc" id="L161">						File file = CasUtil.getFileFor(workingDir, filePath);						</span>
<span class="nc" id="L162">						iteratorLists.add(createIteratorFor(file));					</span>
<span class="nc" id="L163">		        }</span>
<span class="nc" id="L164">				iterators.add(new InterleavedStreamingIterators(iteratorLists));</span>
<span class="nc" id="L165">			} catch (Exception e) {</span>
<span class="nc bnc" id="L166" title="All 2 branches missed.">				for(StreamingIterator&lt;? extends Trace&gt; iter : iteratorLists){</span>
<span class="nc" id="L167">					IOUtil.closeAndIgnoreErrors(iter);</span>
<span class="nc" id="L168">				}</span>
<span class="nc bnc" id="L169" title="All 2 branches missed.">				for(StreamingIterator&lt;? extends Trace&gt; iter : iterators){</span>
<span class="nc" id="L170">					IOUtil.closeAndIgnoreErrors(iter);</span>
<span class="nc" id="L171">				}</span>
<span class="nc" id="L172">				throw new IllegalStateException(&quot;error getting input read data&quot;, e);</span>
<span class="nc" id="L173">			}   </span>
		}
		
<span class="fc" id="L176">	}</span>
	
	private StreamingIterator&lt;? extends Trace&gt; createIteratorFor(File file) throws DataStoreException, IOException{
<span class="fc" id="L179">        ReadFileType readType = ReadFileType.getTypeFromFile(file);</span>
<span class="pc bpc" id="L180" title="3 of 4 branches missed.">           switch(readType){</span>
	            case FASTQ: 
<span class="nc" id="L182">	            	return createFastqIterator(file);</span>
	            case SFF:
<span class="nc" id="L184">	            	return createSffIterator(file);</span>
	            case FASTA:
<span class="fc" id="L186">                       return createFastaIterator(file);</span>
	            default: 
<span class="nc" id="L188">	            	throw new IllegalArgumentException(&quot;unsupported type &quot;+ file.getName());</span>
	            }
        
   }


    private StreamingIterator&lt;? extends Trace&gt; createFastqIterator(File illuminaFile) throws DataStoreException {
		try {
<span class="nc" id="L196">		    FastqFileDataStore datastore = createFastqDataStore(illuminaFile);</span>
<span class="nc" id="L197">			return createIteratorFor(datastore);</span>
<span class="nc" id="L198">		} catch (IOException e) {</span>
<span class="nc" id="L199">			throw new IllegalStateException(&quot;error processing fastq file: &quot;+ illuminaFile.getAbsolutePath(), e);</span>
		}
		
    }



    protected StreamingIterator&lt;? extends Trace&gt; createIteratorFor(FastqFileDataStore datastore) throws DataStoreException {
<span class="nc" id="L207">        return new RemoveWhitespaceFromIdAdapter(datastore.iterator());</span>
    }


    /**
     * Create a new FastqDataStore to process the {@link FastqRecord}s
     * in the given fastq file.  This method can be overwritten to 
     * override the default behavior.
     * 
     * @param illuminaFile the file to parse; will not be null, but may not
     * exist if the cas file is corrupted.
     * 
     * @return a new FastqFileDataStore instance; can not be null.
     * @throws IOException if there is a problem parsing the file.
     * 
     * @since 5.2
     */
    protected FastqFileDataStore createFastqDataStore(File fastqFile)
            throws IOException {
<span class="nc" id="L226">        FastqFileDataStoreBuilder builder = new FastqFileDataStoreBuilder(fastqFile)</span>
<span class="nc" id="L227">        								.hint(DataStoreProviderHint.ITERATION_ONLY);</span>
<span class="nc" id="L228">        FastqQualityCodec codecToUse = getQualityCodec();</span>
<span class="nc bnc" id="L229" title="All 2 branches missed.">        if(codecToUse !=null){</span>
<span class="nc" id="L230">        	builder.qualityCodec(codecToUse);</span>
        }
<span class="nc" id="L232">        return builder.build();</span>
    }

    protected StreamingIterator&lt;? extends Trace&gt; createSffIterator(File sffFile) throws DataStoreException, IOException{
<span class="nc" id="L236">        return new SffFileDataStoreBuilder(sffFile).hint(DataStoreProviderHint.ITERATION_ONLY).build().iterator();</span>
    }

    protected StreamingIterator&lt;? extends Trace&gt; createFastaIterator(File fastaFile) throws DataStoreException{        
        try {
<span class="fc" id="L241">			NucleotideFastaDataStore datastore = new NucleotideFastaFileDataStoreBuilder(fastaFile)</span>
<span class="fc" id="L242">															.hint(DataStoreProviderHint.ITERATION_ONLY)</span>
<span class="fc" id="L243">															.build();</span>
			
			@SuppressWarnings(&quot;unchecked&quot;)
<span class="fc" id="L246">			TraceDataStore&lt;Trace&gt; fakeQualities = DataStore.adapt(TraceDataStore.class, datastore, </span>
					from -&gt;{
<span class="fc" id="L248">						        int numberOfQualities =(int) from.getSequence().getLength();</span>
<span class="fc" id="L249">								byte[] qualities = new byte[numberOfQualities];</span>
<span class="fc" id="L250">								Arrays.fill(qualities, PhredQuality.valueOf(30).getQualityScore());</span>
<span class="fc" id="L251">						        final QualitySequence qualSequence = new QualitySequenceBuilder(qualities).build();</span>
<span class="fc" id="L252">							return new Trace() {</span>
								
								@Override
								public QualitySequence getQualitySequence() {
<span class="nc" id="L256">									return qualSequence;</span>
								}
								
								@Override
								public NucleotideSequence getNucleotideSequence() {

<span class="fc" id="L262">									return from.getSequence();</span>
								}
								
								@Override
								public String getId() {
<span class="fc" id="L267">									return from.getId();</span>
								}
							};
						}
				
			);
<span class="fc" id="L273">			return fakeQualities.iterator();</span>
<span class="nc" id="L274">        } catch (IOException e) {</span>
<span class="nc" id="L275">			throw new DataStoreException(&quot;error reading fasta file &quot;+ fastaFile.getAbsolutePath(),e);</span>
		}
    }
	/**
	 * The given {@link Trace} did not align to any references.    
	 * @param currentTrace the complete {@link Trace} that did not match.
	 */
    protected abstract void notAligned(Trace currentTrace);
    /**
     * The given {@link Trace} aligned to the given reference id
     * with the given {@link CasMatch}.
     * @param referenceId the reference id that the read aligned to.
     * @param match the {@link CasMatch} of this read describing the alignment.
     * 
     * @param traceOfRead the complete {@link Trace} that aligned.
     */
    protected abstract void aligned(Trace traceOfRead, String referenceId, CasMatch match);
    
	@Override
	public CasMatchVisitor visitMatches(CasVisitorCallback callback) {
		
<span class="fc" id="L296">		chainedTraceIterator = IteratorUtil.createChainedStreamingIterator(iterators);</span>
<span class="fc" id="L297">		return new TraceCasMatchVisitor(chainedTraceIterator);</span>
		
	}
	
	private class TraceCasMatchVisitor implements CasMatchVisitor{
		private final StreamingIterator&lt;Trace&gt; chainedTraceIterator;
		
		public TraceCasMatchVisitor(
<span class="fc" id="L305">				StreamingIterator&lt;Trace&gt; chainedTraceIterator) {</span>
<span class="fc" id="L306">			this.chainedTraceIterator = chainedTraceIterator;</span>
<span class="fc" id="L307">		}</span>

		@Override
		public void visitMatch(CasMatch match) {
<span class="pc bpc" id="L311" title="1 of 2 branches missed.">			if(!chainedTraceIterator.hasNext()){</span>
<span class="nc" id="L312">				closeIterator();</span>
<span class="nc" id="L313">				throw new IllegalStateException(&quot;possible cas file corruption : no more reads in input files but cas file says there are more reads&quot;);</span>
			}
<span class="fc" id="L315">			Trace currentTrace = chainedTraceIterator.next();</span>
<span class="fc bfc" id="L316" title="All 2 branches covered.">			if(match.matchReported()){</span>
<span class="fc" id="L317">				CasAlignment alignment = match.getChosenAlignment();</span>
<span class="fc" id="L318">				int refIndex = (int)alignment.getReferenceIndex();</span>
				
<span class="fc" id="L320">				String refId = referenceIds.get(refIndex);</span>
				
<span class="fc" id="L322">				aligned(currentTrace, refId, match);</span>
				
<span class="fc" id="L324">			}else{</span>
<span class="fc" id="L325">				AbstractReadCasVisitor.this.notAligned(currentTrace);</span>
			}
			
<span class="fc" id="L328">		}</span>
		

		

		@Override
		public void visitEnd() {
<span class="fc" id="L335">			closeIterator();</span>
<span class="fc" id="L336">		}</span>

		@Override
		public void halted() {
<span class="nc" id="L340">			closeIterator();</span>
<span class="nc" id="L341">		}</span>
		
		
	}
	
	protected void closeIterator(){
<span class="fc" id="L347">		IOUtil.closeAndIgnoreErrors(chainedTraceIterator);</span>
<span class="fc" id="L348">	}</span>
	
	private static class RemoveWhitespaceFromIdAdapter implements StreamingIterator&lt;FastqRecord&gt;{
		private final  StreamingIterator&lt;FastqRecord&gt; delegate;

<span class="nc" id="L353">		private static Pattern WHITESPACE_PATTERN = Pattern.compile(&quot;^(\\S+)\\s+(\\S+)$&quot;);</span>
<span class="nc" id="L354">		public RemoveWhitespaceFromIdAdapter(StreamingIterator&lt;FastqRecord&gt; delegate) {</span>
<span class="nc" id="L355">			this.delegate = delegate;</span>
<span class="nc" id="L356">		}</span>

		@Override
		public boolean hasNext() {
<span class="nc" id="L360">			return delegate.hasNext();</span>
		}

		@Override
		public void close() {
<span class="nc" id="L365">			delegate.close();</span>
			
<span class="nc" id="L367">		}</span>

		@Override
		public FastqRecord next() {
<span class="nc" id="L371">			FastqRecord record =delegate.next();</span>
<span class="nc" id="L372">			Matcher matcher= WHITESPACE_PATTERN.matcher(record.getId());</span>
<span class="nc bnc" id="L373" title="All 2 branches missed.">			if(matcher.matches()){</span>
<span class="nc" id="L374">			        return record.toBuilder()</span>
<span class="nc" id="L375">			                        .id(matcher.group(1)+&quot;_&quot;+matcher.group(2))</span>
<span class="nc" id="L376">			                        .build();</span>
				
			}
<span class="nc" id="L379">			return record;</span>
		}

		@Override
		public void remove() {
<span class="nc" id="L384">			delegate.remove();</span>
			
<span class="nc" id="L386">		}	</span>
	}
	
	private static class InterleavedStreamingIterators implements StreamingIterator&lt;Trace&gt;{
		private final List&lt;StreamingIterator&lt;? extends Trace&gt;&gt; iterators;

<span class="nc" id="L392">		private int offset=0;</span>
		
		public InterleavedStreamingIterators(
<span class="nc" id="L395">				List&lt;StreamingIterator&lt;? extends Trace&gt;&gt; iterators) {</span>
<span class="nc" id="L396">			this.iterators = iterators;</span>
<span class="nc" id="L397">		}</span>

		@Override
		public boolean hasNext() {
<span class="nc" id="L401">			return iterators.get(offset).hasNext();</span>
		}

		@Override
		public void close() {
<span class="nc bnc" id="L406" title="All 2 branches missed.">			for(StreamingIterator&lt;? extends Trace&gt; iter : iterators){</span>
<span class="nc" id="L407">				IOUtil.closeAndIgnoreErrors(iter);</span>
<span class="nc" id="L408">			}			</span>
			
<span class="nc" id="L410">		}</span>

		@Override
		public Trace next() {
<span class="nc" id="L414">			Trace next = iterators.get(offset++).next();</span>
<span class="nc bnc" id="L415" title="All 2 branches missed.">			if(offset ==iterators.size()){</span>
<span class="nc" id="L416">				offset=0;</span>
			}
<span class="nc" id="L418">			return next;</span>
		}

		@Override
		public void remove() {
<span class="nc" id="L423">			throw new UnsupportedOperationException();</span>
			
		}
		
		
	}
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.5.201505241946</span></div></body></html>