<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../.resources/report.css" type="text/css"/><link rel="shortcut icon" href="../.resources/report.gif" type="image/gif"/><title>IupacTranslationTables.java</title><link rel="stylesheet" href="../.resources/prettify.css" type="text/css"/><script type="text/javascript" src="../.resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../.sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">Jillion</a> &gt; <a href="index.source.html" class="el_package">org.jcvi.jillion.core.residue.aa</a> &gt; <span class="el_source">IupacTranslationTables.java</span></div><h1>IupacTranslationTables.java</h1><pre class="source lang-java linenums">/*******************************************************************************
 * Jillion development code
 * 
 * This code may be freely distributed and modified under the
 * terms of the GNU Lesser General Public License.  This should
 * be distributed with the code.  If you do not have a copy,
 *  see:
 * 
 *          http://www.gnu.org/copyleft/lesser.html
 * 
 * 
 * Copyright for this code is held jointly by the individual authors.  These should be listed in the @author doc comments.
 * 
 * Information about Jillion can be found on its homepage
 * 
 *         http://jillion.sourceforge.net
 * 
 * Contributors:
 *     Danny Katzel - initial API and implementation
 ******************************************************************************/
package org.jcvi.jillion.core.residue.aa;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import org.jcvi.jillion.core.residue.Frame;
import org.jcvi.jillion.core.residue.aa.TranslationVisitor.FoundStartResult;
import org.jcvi.jillion.core.residue.aa.TranslationVisitor.FoundStopResult;
import org.jcvi.jillion.core.residue.nt.NucleotideSequence;
import org.jcvi.jillion.core.residue.nt.Triplet;
import org.jcvi.jillion.core.util.MapUtil;

<span class="pc" id="L34">public enum IupacTranslationTables implements TranslationTable{</span>
	
	
	
	
	/**
	 * Default translation table used by GenBank.
	 */
<span class="fc" id="L42">	STANDARD(1),</span>
<span class="fc" id="L43">	VERTEBRATE_MITOCHONDRIAL(2){</span>

		@Override
		protected void updateTable(Map&lt;Triplet, Codon&gt; map) {
<span class="fc" id="L47">			insertIntoTable('A', 'G', 'A', AminoAcid.STOP);</span>
<span class="fc" id="L48">			insertIntoTable('A', 'G', 'G', AminoAcid.STOP);</span>
<span class="fc" id="L49">			insertIntoTable('A', 'G', 'R', AminoAcid.STOP);</span>
			
<span class="fc" id="L51">			insertIntoTable('A', 'T', 'A', AminoAcid.Methionine, true);</span>
<span class="fc" id="L52">			insertIntoTable('A', 'T', 'R', AminoAcid.Methionine, true);</span>
			
<span class="fc" id="L54">			insertIntoTable('T', 'G', 'A', AminoAcid.Tryptophan);</span>
<span class="fc" id="L55">			insertIntoTable('T', 'G', 'R', AminoAcid.Tryptophan);</span>
<span class="fc" id="L56">		}</span>
		
	},
<span class="fc" id="L59">	YEAST_MITOCHONDRIAL(3){</span>

		@Override
		protected void updateTable(Map&lt;Triplet, Codon&gt; map) {			
			
<span class="fc" id="L64">			insertIntoTable('A', 'T', 'A', AminoAcid.Methionine, true);</span>
<span class="fc" id="L65">			insertIntoTable('A', 'T', 'R', AminoAcid.Methionine, true);</span>
			
<span class="fc" id="L67">			insertIntoTable('C', 'T', 'A', AminoAcid.Threonine);</span>
<span class="fc" id="L68">			insertIntoTable('C', 'T', 'C', AminoAcid.Threonine);</span>
<span class="fc" id="L69">			insertIntoTable('C', 'T', 'G', AminoAcid.Threonine);</span>
<span class="fc" id="L70">			insertIntoTable('C', 'T', 'T', AminoAcid.Threonine);</span>
			
			
<span class="fc" id="L73">			insertIntoTable('C', 'T', 'M', AminoAcid.Threonine);</span>
<span class="fc" id="L74">			insertIntoTable('C', 'T', 'R', AminoAcid.Threonine);</span>
<span class="fc" id="L75">			insertIntoTable('C', 'T', 'W', AminoAcid.Threonine);</span>
<span class="fc" id="L76">			insertIntoTable('C', 'T', 'S', AminoAcid.Threonine);</span>
<span class="fc" id="L77">			insertIntoTable('C', 'T', 'Y', AminoAcid.Threonine);</span>
<span class="fc" id="L78">			insertIntoTable('C', 'T', 'K', AminoAcid.Threonine);</span>
<span class="fc" id="L79">			insertIntoTable('C', 'T', 'V', AminoAcid.Threonine);</span>
<span class="fc" id="L80">			insertIntoTable('C', 'T', 'H', AminoAcid.Threonine);</span>
<span class="fc" id="L81">			insertIntoTable('C', 'T', 'D', AminoAcid.Threonine);</span>
<span class="fc" id="L82">			insertIntoTable('C', 'T', 'B', AminoAcid.Threonine);</span>
<span class="fc" id="L83">			insertIntoTable('C', 'T', 'N', AminoAcid.Threonine);</span>
			
<span class="fc" id="L85">			insertIntoTable('T', 'G', 'A', AminoAcid.Tryptophan);</span>
			
<span class="fc" id="L87">			removeFromTable('C', 'G', 'A');</span>
<span class="fc" id="L88">			removeFromTable('C', 'G', 'C');</span>
<span class="fc" id="L89">			removeFromTable('C', 'G', 'R');</span>
			

<span class="fc" id="L92">		}</span>
		
	},

<span class="fc" id="L96">	MOLD_PROTOZOAN_COELENTERATE_MITOCHONDRIAL_AND_MYCOPLASMA_SPIROPLAMSA(4){</span>

		@Override
		protected void updateTable(Map&lt;Triplet, Codon&gt; map) {
<span class="fc" id="L100">			insertIntoTable('T', 'G', 'A', AminoAcid.Tryptophan);</span>
<span class="fc" id="L101">			insertIntoTable('T', 'G', 'R', AminoAcid.Tryptophan);</span>
			
<span class="fc" id="L103">		}</span>
		
	},
	/**
	 * Nucleotide Translation table used by
	 * &lt;ul&gt;
	 * &lt;li&gt;Ciliata - Oxytricha and Stylonycha, Paramecium, Tetrahymena, Oxytrichidae etc. &lt;/li&gt;
	 * &lt;li&gt;Dasycladaceae - Acetabulara and Batophora&lt;/li&gt;
	 * &lt;li&gt;Diplomonadida - Scope: Hexamita&lt;/li&gt;
	 * &lt;/ul&gt;
	 */
<span class="fc" id="L114">	CILIATE_DASYCLADACEAN_AND_HEXAMITA(6){</span>

		@Override
		protected void updateTable(Map&lt;Triplet, Codon&gt; map) {
<span class="fc" id="L118">			insertIntoTable('T', 'A', 'A', AminoAcid.Glutamine);</span>
<span class="fc" id="L119">			insertIntoTable('T', 'A', 'G', AminoAcid.Glutamine);</span>
<span class="fc" id="L120">			insertIntoTable('T', 'A', 'R', AminoAcid.Glutamine);</span>
			
			//not a start codons anymore
<span class="fc" id="L123">			insertIntoTable('T', 'T', 'G', AminoAcid.Leucine);</span>
<span class="fc" id="L124">			insertIntoTable('C', 'T', 'G', AminoAcid.Leucine);</span>
<span class="fc" id="L125">		}</span>
		
	},
<span class="fc" id="L128">	BACTERIAL_ARCHAEL_AND_PLANT_PLASTID(11){</span>

		@Override
		protected void updateTable(Map&lt;Triplet, Codon&gt; map) {
			//more start codons only?
<span class="fc" id="L133">			insertIntoTable('G', 'T', 'G', AminoAcid.Valine, true);</span>
			
<span class="fc" id="L135">			insertIntoTable('A', 'T', 'T', AminoAcid.Isoleucine, true);</span>
<span class="fc" id="L136">			insertIntoTable('A', 'T', 'C', AminoAcid.Isoleucine, true);</span>
<span class="fc" id="L137">			insertIntoTable('A', 'T', 'A', AminoAcid.Isoleucine, true);</span>
<span class="fc" id="L138">			insertIntoTable('A', 'T', 'G', AminoAcid.Isoleucine, true);</span>
			
<span class="fc" id="L140">			insertIntoTable('A', 'T', 'M', AminoAcid.Isoleucine, true);</span>
<span class="fc" id="L141">			insertIntoTable('A', 'T', 'R', AminoAcid.Isoleucine, true);</span>
<span class="fc" id="L142">			insertIntoTable('A', 'T', 'W', AminoAcid.Isoleucine, true);</span>
<span class="fc" id="L143">			insertIntoTable('A', 'T', 'S', AminoAcid.Isoleucine, true);</span>
<span class="fc" id="L144">			insertIntoTable('A', 'T', 'Y', AminoAcid.Isoleucine, true);</span>
<span class="fc" id="L145">			insertIntoTable('A', 'T', 'K', AminoAcid.Isoleucine, true);</span>
<span class="fc" id="L146">			insertIntoTable('A', 'T', 'V', AminoAcid.Isoleucine, true);</span>
<span class="fc" id="L147">			insertIntoTable('A', 'T', 'H', AminoAcid.Isoleucine, true);</span>
<span class="fc" id="L148">			insertIntoTable('A', 'T', 'D', AminoAcid.Isoleucine, true);</span>
<span class="fc" id="L149">			insertIntoTable('A', 'T', 'B', AminoAcid.Isoleucine, true);</span>
<span class="fc" id="L150">			insertIntoTable('A', 'T', 'N', AminoAcid.Isoleucine, true);</span>
<span class="fc" id="L151">		}</span>
		
	},
	
	;

	private static final Map&lt;Integer, IupacTranslationTables&gt; TABLES;
	static{
<span class="fc" id="L159">		TABLES = new HashMap&lt;Integer, IupacTranslationTables&gt;(MapUtil.computeMinHashMapSizeWithoutRehashing(25));</span>
<span class="fc bfc" id="L160" title="All 2 branches covered.">		for(IupacTranslationTables table : values()){</span>
<span class="fc" id="L161">			TABLES.put(Integer.valueOf(table.getTableNumber()), table);</span>
		}
<span class="fc" id="L163">	}</span>
	
<span class="fc" id="L165">	private final Map&lt;Triplet, Codon&gt; map = new HashMap&lt;Triplet, Codon&gt;(MapUtil.computeMinHashMapSizeWithoutRehashing(200));</span>
	private final byte tableNumber;
	
<span class="fc" id="L168">	private IupacTranslationTables(int tableNumber){</span>
<span class="fc" id="L169">		initialzeTable();</span>
<span class="fc" id="L170">		updateTable(map);</span>
<span class="fc" id="L171">		this.tableNumber = (byte)tableNumber;</span>
		
<span class="fc" id="L173">	}</span>
	
	protected void removeFromTable(char base1, char base2, char base3){
<span class="fc" id="L176">		Triplet triplet = Triplet.create(base1, base2, base3);</span>
<span class="fc" id="L177">		map.remove(triplet);</span>
<span class="fc" id="L178">	}</span>
	protected void  insertIntoTable(char base1, char base2, char base3, AminoAcid aa){
<span class="fc" id="L180">		insertIntoTable(base1, base2, base3, aa, false);</span>
<span class="fc" id="L181">	}</span>
	protected void  insertIntoTable(char base1, char base2, char base3, AminoAcid aa, boolean isStart){
<span class="fc" id="L183">		Triplet triplet = Triplet.create(base1, base2, base3);</span>
<span class="fc" id="L184">		Codon.Builder builder = new Codon.Builder(triplet, aa);</span>
<span class="fc bfc" id="L185" title="All 2 branches covered.">		if(aa == AminoAcid.STOP){</span>
<span class="fc" id="L186">			builder.isStop(true);</span>
		}
<span class="fc bfc" id="L188" title="All 2 branches covered.">		if(isStart){</span>
<span class="fc" id="L189">			builder.isStart(true);</span>
		}
<span class="fc" id="L191">		map.put(triplet, builder.build());</span>
<span class="fc" id="L192">	}</span>
	
	private void initialzeTable(){
		/*
		 * This group of Strings
		 * explains the amino acid,
		 *  if it's a start codon (M)
		 * and the triplet used to encode it.
		 * 
		 * This way of displaying the translation matrix
		 * is used by NCBI Genetic Codes page
		 * http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi
		 * 
		 * I have made some changes, mostly adding ambiguity support
		 * for wobble bases.
		 * 
		 * Non-standard translation tables will update the
		 * Map later.
		 */
		
<span class="fc" id="L212">		char[] aas =    &quot;FFFLLLSSSSSSSSSSSSSSSYYY***CCC*WLLLLLLLLPPPPPPPPPPPPPPPHHHQQQRRRRRRRRRRRRRRRIIIIIIIMTTTTTTTTTTTTTTTNNNKKKSSSRRRVVVVVVVVVVVVVVVAAAAAAAAAAAAAAADDDEEEGGGGGGGGGGGGGGG&quot;.toCharArray();</span>
<span class="fc" id="L213">		char[] starts = &quot;----M----------------------------------M-------------------------------------------M------------------------------------------------------------------------------&quot;.toCharArray();</span>
<span class="fc" id="L214">		char[] base1 =  &quot;TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG&quot;.toCharArray();</span>
<span class="fc" id="L215">		char[] base2 =  &quot;TTTTTTCCCCCCCCCCCCCCCAAAAAAGGGGGTTTTTTTTCCCCCCCCCCCCCCCAAAAAAGGGGGGGGGGGGGGGTTTTTTTTCCCCCCCCCCCCCCCAAAAAAGGGGGGTTTTTTTTTTTTTTTCCCCCCCCCCCCCCCAAAAAAGGGGGGGGGGGGGGG&quot;.toCharArray();</span>
<span class="fc" id="L216">		char[] base3 =  &quot;TCYAGRTCAGMRWSYKVHDBNTCYAGRTCYAGTCAMWYHGTCAGMRWSYKVHDBNTCYAGRTCAGMRWSYKVHDBNTCAMWYHGTCAGMRWSYKVHDBNTCYAGRTCYAGRTCAGMRWSYKVHDBNTCAGMRWSYKVHDBNTCYAGRTCAGMRWSYKVHDBN&quot;.toCharArray();			</span>
		
<span class="fc bfc" id="L218" title="All 2 branches covered.">		for(int i=0; i&lt;aas.length; i++){</span>
<span class="fc" id="L219">			insertIntoTable(base1[i], base2[i], base3[i], </span>
<span class="fc bfc" id="L220" title="All 2 branches covered.">								AminoAcid.parse(aas[i]),</span>
								starts[i] =='M');
			
		}
		
		//add gap
<span class="fc" id="L226">		insertIntoTable('-','-','-', AminoAcid.Gap, false);</span>
<span class="fc" id="L227">	}</span>
	

	@Override
	public ProteinSequence translate(NucleotideSequence sequence) {
<span class="fc" id="L232">		return translate(sequence, Frame.ONE,true);</span>
	}
	@Override
	public ProteinSequence translate(NucleotideSequence sequence, boolean substituteStart) {
<span class="fc" id="L236">		return translate(sequence, Frame.ONE, substituteStart);</span>
	}

	@Override
	public ProteinSequence translate(NucleotideSequence sequence, Frame frame) {
<span class="fc" id="L241">		return translate(sequence, frame, true);</span>
	}
	@Override
	public ProteinSequence translate(NucleotideSequence sequence, Frame frame, boolean substituteStart) {
<span class="fc" id="L245">		return translate(sequence, frame, (int)sequence.getLength(), substituteStart);</span>
	}
	@Override
	public ProteinSequence translate(NucleotideSequence sequence, Frame frame, int length, boolean substituteStart) {
<span class="pc bpc" id="L249" title="1 of 2 branches missed.">		if(sequence ==null){</span>
<span class="nc" id="L250">			throw new NullPointerException(&quot;sequence can not be null&quot;);</span>
		}
<span class="pc bpc" id="L252" title="1 of 2 branches missed.">		if(frame ==null){</span>
<span class="nc" id="L253">			throw new NullPointerException(&quot;frame can not be null&quot;);</span>
		}
		//Brian says legacy PFGRC and (BioJava and BioPerl?)
		//don't correctly handle the 'not first starts'
		//so if translation table says codon is a start
		//and we've already seen a start, then make it not the start?

<span class="fc" id="L260">		ProteinSequenceBuilder builder = new ProteinSequenceBuilder(length/3);</span>
		
<span class="fc" id="L262">		Iterator&lt;Triplet&gt; iter = frame.asTriplets(sequence);</span>
		
<span class="fc bfc" id="L264" title="All 2 branches covered.">		boolean seenStart=!substituteStart;</span>
<span class="fc" id="L265">		long currentOffset=0;</span>
		
<span class="pc bpc" id="L267" title="1 of 4 branches missed.">		while(iter.hasNext() &amp;&amp; currentOffset &lt;length){</span>
<span class="fc" id="L268">			Triplet triplet =iter.next();</span>
<span class="fc" id="L269">			currentOffset+=3;</span>
<span class="pc bpc" id="L270" title="1 of 2 branches missed.">			if(triplet !=null){</span>
<span class="fc" id="L271">				Codon codon =translate(triplet);</span>
<span class="fc bfc" id="L272" title="All 4 branches covered.">				if(codon.isStart() &amp;&amp; !seenStart){</span>
<span class="fc" id="L273">					seenStart=true;</span>
					//hardcode an M if this is our first start
					//which may 
					//not be the amino acid returned by 
					//#getAminoAcid() depending on the translation table
<span class="fc" id="L278">					builder.append(AminoAcid.Methionine);</span>
				}else{
<span class="fc" id="L280">					builder.append(codon.getAminoAcid());</span>
				}
			}
<span class="fc" id="L283">		}</span>
<span class="fc" id="L284">		return builder.build();</span>
	}
	
	
	@Override
        public void translate(NucleotideSequence sequence, Frame frame, TranslationVisitor visitor) {
<span class="pc bpc" id="L290" title="1 of 2 branches missed.">                if(sequence ==null){</span>
<span class="nc" id="L291">                        throw new NullPointerException(&quot;sequence can not be null&quot;);</span>
                }
<span class="pc bpc" id="L293" title="1 of 2 branches missed.">                if(frame ==null){</span>
<span class="nc" id="L294">                        throw new NullPointerException(&quot;frame can not be null&quot;);</span>
                }
<span class="pc bpc" id="L296" title="1 of 2 branches missed.">                if(visitor ==null){</span>
<span class="nc" id="L297">                    throw new NullPointerException(&quot;frame can not be null&quot;);</span>
                }
                //Brian says legacy PFGRC and (BioJava and BioPerl?)
                //don't correctly handle the 'not first starts'
                //so if translation table says codon is a start
                //and we've already seen a start, then make it not the start?
<span class="fc" id="L303">                Iterator&lt;Triplet&gt; iter = frame.asTriplets(sequence);</span>
<span class="fc" id="L304">                boolean seenStart=false;</span>
<span class="fc" id="L305">                long currentOffset=frame.ordinal();</span>
               
<span class="fc bfc" id="L307" title="All 2 branches covered.">                while(iter.hasNext()){</span>
<span class="fc" id="L308">                        Triplet triplet =iter.next();</span>
                        
<span class="pc bpc" id="L310" title="1 of 2 branches missed.">                        if(triplet !=null){</span>
<span class="fc" id="L311">                                Codon codon =translate(triplet);</span>
<span class="fc bfc" id="L312" title="All 2 branches covered.">                                if(codon.isStart()){</span>
<span class="fc" id="L313">                                    FoundStartResult result = visitor.foundStart(currentOffset, codon);</span>
<span class="pc bpc" id="L314" title="1 of 2 branches missed.">                                    if(result ==FoundStartResult.STOP){</span>
<span class="nc" id="L315">                                        break;</span>
                                    }
<span class="pc bpc" id="L317" title="1 of 2 branches missed.">                                    seenStart = result != FoundStartResult.FIND_ADDITIONAL_STARTS;</span>
<span class="fc bfc" id="L318" title="All 2 branches covered.">                                }else if(seenStart){</span>
<span class="fc bfc" id="L319" title="All 2 branches covered.">                                    if(codon.isStop()){</span>
<span class="fc" id="L320">                                        FoundStopResult result = visitor.foundStop(currentOffset, codon);</span>
<span class="pc bpc" id="L321" title="1 of 2 branches missed.">                                        if(result == FoundStopResult.STOP){</span>
<span class="fc" id="L322">                                            break;</span>
                                        }
                                        
<span class="nc" id="L325">                                    }else{</span>
<span class="fc" id="L326">                                        visitor.visitCodon(currentOffset, codon);</span>
                                    }
                                }
                        }
<span class="fc" id="L330">                        currentOffset+=3;</span>
<span class="fc" id="L331">                }</span>
<span class="fc" id="L332">                visitor.end();</span>
<span class="fc" id="L333">        }</span>
	@Override
	public ProteinSequence translate(NucleotideSequence sequence, Frame frame, int length) {
<span class="nc" id="L336">		return translate(sequence, frame, length,true);</span>
	}

	
	
	

	protected void updateTable(Map&lt;Triplet, Codon&gt; map){
		//no-op
<span class="fc" id="L345">	}</span>
	private Codon translate(Triplet triplet){
<span class="fc" id="L347">	    return  map.computeIfAbsent(triplet, </span>
<span class="nc" id="L348">		        t -&gt; new Codon.Builder(t, AminoAcid.Unknown_Amino_Acid).build());</span>
		
	}

	public int getTableNumber(){
<span class="fc" id="L353">		return tableNumber;</span>
	}
	
	public static TranslationTable getTableByTableNumber(int tableNumber){
<span class="fc" id="L357">		TranslationTable table = TABLES.get(tableNumber);</span>
<span class="pc bpc" id="L358" title="1 of 2 branches missed.">		if(table ==null){</span>
<span class="nc" id="L359">			throw new IllegalArgumentException(&quot;unknown table number &quot;+ tableNumber);</span>
		}
<span class="fc" id="L361">		return table;</span>
	}
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.5.201505241946</span></div></body></html>