<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../.resources/report.css" type="text/css"/><link rel="shortcut icon" href="../.resources/report.gif" type="image/gif"/><title>OrfFinder.java</title><link rel="stylesheet" href="../.resources/prettify.css" type="text/css"/><script type="text/javascript" src="../.resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../.sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">Jillion</a> &gt; <a href="index.source.html" class="el_package">org.jcvi.jillion.orf</a> &gt; <span class="el_source">OrfFinder.java</span></div><h1>OrfFinder.java</h1><pre class="source lang-java linenums">/*******************************************************************************
 * Jillion development code
 * 
 * This code may be freely distributed and modified under the
 * terms of the GNU Lesser General Public Licence.  This should
 * be distributed with the code.  If you do not have a copy,
 *  see:
 * 
 *          http://www.gnu.org/copyleft/lesser.html
 * 
 * 
 * Copyright for this code is held jointly by the individual authors.  These should be listed in the @author doc comments.
 * 
 * Information about Jillion can be found on its homepage
 * 
 *         http://jillion.sourceforge.net
 * 
 * Contributors:
 *     Danny Katzel - initial API and implementation
 ******************************************************************************/
package org.jcvi.jillion.orf;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;

import org.jcvi.jillion.core.Range;
import org.jcvi.jillion.core.residue.Frame;
import org.jcvi.jillion.core.residue.aa.AminoAcid;
import org.jcvi.jillion.core.residue.aa.Codon;
import org.jcvi.jillion.core.residue.aa.IupacTranslationTables;
import org.jcvi.jillion.core.residue.aa.ProteinSequenceBuilder;
import org.jcvi.jillion.core.residue.aa.TranslationTable;
import org.jcvi.jillion.core.residue.aa.TranslationVisitor;
import org.jcvi.jillion.core.residue.nt.NucleotideSequence;

<span class="fc" id="L39">public class OrfFinder {</span>
    
<span class="pc" id="L41">    public enum FinderOptions{</span>
<span class="fc" id="L42">        SEARCH_FORWARD,</span>
<span class="fc" id="L43">        SEARCH_REVERSE</span>
    }
    
    public List&lt;Orf&gt; find(NucleotideSequence seq){
<span class="fc" id="L47">        return find(seq, IupacTranslationTables.STANDARD,FinderOptions.SEARCH_FORWARD, FinderOptions.SEARCH_REVERSE);</span>
    }
    public List&lt;Orf&gt; find(NucleotideSequence seq, TranslationTable translationTable, FinderOptions...finderOptions){
<span class="fc" id="L50">        List&lt;Orf&gt; orfs = new ArrayList&lt;&gt;();</span>
        
<span class="fc" id="L52">        Set&lt;FinderOptions&gt; options = EnumSet.noneOf(FinderOptions.class);</span>
<span class="fc bfc" id="L53" title="All 2 branches covered.">        for(FinderOptions o : finderOptions){</span>
<span class="fc" id="L54">            options.add(o);</span>
        }
        
<span class="pc bpc" id="L57" title="1 of 2 branches missed.">        if(options.contains(FinderOptions.SEARCH_FORWARD)){</span>
<span class="fc bfc" id="L58" title="All 2 branches covered.">            for(Frame f : Frame.forwardFrames()){</span>
<span class="fc" id="L59">                OrfVisitor visitor = new OrfVisitor(f);</span>
<span class="fc" id="L60">                translationTable.translate(seq, f, visitor);</span>
<span class="fc" id="L61">                visitor.getOrf().ifPresent(orf -&gt; orfs.add(orf));</span>
<span class="fc" id="L62">            }</span>
        }
<span class="pc bpc" id="L64" title="1 of 2 branches missed.">        if(options.contains(FinderOptions.SEARCH_REVERSE)){</span>
<span class="fc" id="L65">            NucleotideSequence reverseSeq = seq.toBuilder().reverseComplement().build();</span>
<span class="fc bfc" id="L66" title="All 2 branches covered.">            for(Frame f : Frame.reverseFrames()){</span>
<span class="fc" id="L67">                OrfVisitor visitor = new OrfVisitor(f);</span>
<span class="fc" id="L68">                translationTable.translate(reverseSeq, f.getOppositeFrame(), visitor);</span>
<span class="pc" id="L69">                visitor.getOrf().ifPresent(orf -&gt; orfs.add(orf));</span>
<span class="fc" id="L70">            }</span>
        }
        
<span class="fc" id="L73">        return orfs;</span>
    }
    
    private static class OrfVisitor implements TranslationVisitor{

<span class="fc" id="L78">        ProteinSequenceBuilder builder = new ProteinSequenceBuilder();</span>
        Frame frame;
        
<span class="fc" id="L81">        public OrfVisitor(Frame frame) {</span>
<span class="fc" id="L82">            this.frame = frame;</span>
<span class="fc" id="L83">        }</span>

<span class="fc" id="L85">        long startCoord =-1; long stopCoord =-1;</span>
<span class="fc" id="L86">        boolean hasStart=false;</span>
<span class="fc" id="L87">        boolean hasStop=false;</span>
        
<span class="fc" id="L89">        Orf orf=null;</span>
        @Override
        public void visitCodon(long nucleotideCoordinate, Codon codon) {
<span class="fc" id="L92">            builder.append(codon.getAminoAcid());</span>
            
<span class="fc" id="L94">        }</span>

        @Override
        public FoundStartResult foundStart(long nucleotideCoordinate,
                Codon codon) {
<span class="fc bfc" id="L99" title="All 2 branches covered.">            if(hasStart){</span>
                //already seen a start
<span class="fc" id="L101">                builder.append(codon.getAminoAcid());</span>
            }else{
<span class="fc" id="L103">            hasStart=true;</span>
<span class="fc" id="L104">            startCoord = nucleotideCoordinate;</span>
          //hardcode an M if this is our first start
            //which may 
            //not be the amino acid returned by 
            //#getAminoAcid() depending on the translation table
<span class="fc" id="L109">            builder.append(AminoAcid.Methionine);</span>
            }
<span class="fc" id="L111">            return FoundStartResult.CONTINUE;</span>
        }

        @Override
        public FoundStopResult foundStop(long nucleotideCoordinate,
                Codon codon) {
<span class="fc" id="L117">            hasStop = true;</span>
<span class="fc" id="L118">            stopCoord = nucleotideCoordinate+=2; // only +2 to include last base in aa</span>
<span class="fc" id="L119">            builder.append(codon.getAminoAcid());</span>
<span class="fc" id="L120">            return FoundStopResult.STOP;</span>
        }

        @Override
        public void end() {
<span class="fc bfc" id="L125" title="All 4 branches covered.">            if(hasStart &amp;&amp; hasStop){</span>
<span class="fc" id="L126">                orf = new Orf(frame, builder.build(), Range.of(startCoord, stopCoord));</span>
            }
            
<span class="fc" id="L129">        }</span>
        
        public Optional&lt;Orf&gt; getOrf(){
<span class="fc" id="L132">            return Optional.ofNullable(orf);</span>
        }
        
    }
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.5.201505241946</span></div></body></html>