<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../.resources/report.css" type="text/css"/><link rel="shortcut icon" href="../.resources/report.gif" type="image/gif"/><title>PaddedSamTransformationService.java</title><link rel="stylesheet" href="../.resources/prettify.css" type="text/css"/><script type="text/javascript" src="../.resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../.sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">Jillion</a> &gt; <a href="index.source.html" class="el_package">org.jcvi.jillion.sam.transform</a> &gt; <span class="el_source">PaddedSamTransformationService.java</span></div><h1>PaddedSamTransformationService.java</h1><pre class="source lang-java linenums">/*******************************************************************************
 * Jillion development code
 * 
 * This code may be freely distributed and modified under the
 * terms of the GNU Lesser General Public License.  This should
 * be distributed with the code.  If you do not have a copy,
 *  see:
 * 
 *          http://www.gnu.org/copyleft/lesser.html
 * 
 * 
 * Copyright for this code is held jointly by the individual authors.  These should be listed in the @author doc comments.
 * 
 * Information about Jillion can be found on its homepage
 * 
 *         http://jillion.sourceforge.net
 * 
 * Contributors:
 *     Danny Katzel - initial API and implementation
 ******************************************************************************/
package org.jcvi.jillion.sam.transform;

import java.io.File;
import java.io.IOException;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import org.jcvi.jillion.assembly.AssemblyTransformationService;
import org.jcvi.jillion.assembly.AssemblyTransformer;
import org.jcvi.jillion.assembly.AssemblyUtil;
import org.jcvi.jillion.assembly.ReadInfo;
import org.jcvi.jillion.core.Direction;
import org.jcvi.jillion.core.Range;
import org.jcvi.jillion.core.datastore.DataStoreException;
import org.jcvi.jillion.core.qual.QualitySequence;
import org.jcvi.jillion.core.qual.QualitySequenceBuilder;
import org.jcvi.jillion.core.residue.nt.NucleotideSequence;
import org.jcvi.jillion.core.residue.nt.NucleotideSequenceBuilder;
import org.jcvi.jillion.core.util.MapUtil;
import org.jcvi.jillion.internal.core.util.GrowableIntArray;
import org.jcvi.jillion.sam.SamParser;
import org.jcvi.jillion.sam.SamParserFactory;
import org.jcvi.jillion.sam.SamRecord;
import org.jcvi.jillion.sam.SamRecordFlag;
import org.jcvi.jillion.sam.SamRecordFlags;
import org.jcvi.jillion.sam.SamVisitor;
import org.jcvi.jillion.sam.VirtualFileOffset;
import org.jcvi.jillion.sam.cigar.Cigar;
import org.jcvi.jillion.sam.cigar.Cigar.ClipType;
import org.jcvi.jillion.sam.header.SamHeader;
import org.jcvi.jillion.sam.header.SamReferenceSequence;
/**
 * {@code PaddedSamTransformationService}
 * is a class that can parse a &lt;em&gt;padded&lt;/em&gt;SAM file
 * was produced from a denovo assembler
 * and contains the padded read alignments
 * to a padded reference/consensus
 * and call the appropriate methods
 * on a given {@link AssemblyTransformer}
 * so the transformer can get assembly and alignment 
 * information from the padded SAM file without knowing
 * anything about how SAM files are formatted.
 * 
 * @author dkatzel
 *
 */
public final class PaddedSamTransformationService implements AssemblyTransformationService{

	
	private final SamParser parser;
	
	/**
	 * Create a new {@link SamTransformationService} using
	 * the given padded SAM encoded file  which
	 * should already include the gapped reference sequences
	 * inside it.
	 * @param paddedSamOrBamFile the SAM file to parse and transform; can not be null
	 * and must exist.
	 * 
	 * @throws IOException if there is a problem parsing the input file.
	 * @throws NullPointerException if the file is null.
	 */
<span class="fc" id="L86">	public PaddedSamTransformationService(File paddedSamOrBamFile) throws IOException {</span>
		
<span class="fc" id="L88">		parser = SamParserFactory.create(paddedSamOrBamFile);</span>
		
<span class="fc" id="L90">	}</span>
	/**
	 * Parse the SAM file and call the appropriate methods on the given
	 * {@link AssemblyTransformer}.
	 * @param transformer the {@link AssemblyTransformer} instance to call
	 * the methods on; can not be null.
	 * @throws NullPointerException if transformer is null.
	 * @throws IllegalStateException if there is any problem parsing the file
	 * including not providing a padded reference
	 * and not encountering the padded reference 
	 * sequences before they are used in a read alignment.
	 */
	@Override
	public void transform(final AssemblyTransformer transformer){
<span class="pc bpc" id="L104" title="1 of 2 branches missed.">		if(transformer ==null){</span>
<span class="nc" id="L105">			throw new NullPointerException(&quot;transformer can not be null&quot;);</span>
		}
		try {
<span class="fc" id="L108">			SamTransformerVisitor visitor = new SamTransformerVisitor(transformer);</span>
<span class="fc" id="L109">			parser.parse(visitor);</span>
<span class="nc" id="L110">		} catch (Exception e) {</span>
<span class="nc" id="L111">			throw new IllegalStateException(&quot;error parsing sam file&quot;, e);</span>
<span class="fc" id="L112">		}</span>
<span class="fc" id="L113">	}</span>
	
	
	private static final class SamTransformerVisitor implements SamVisitor{

		private final AssemblyTransformer transformer;
		private Map&lt;String,GrowableIntArray&gt; gapOffsetMap;
		private Map&lt;String, NucleotideSequence&gt; paddedReferenceSequences;
		private Set&lt;String&gt; referenceNames;
		
		
<span class="fc" id="L124">		public SamTransformerVisitor(AssemblyTransformer transformer) throws DataStoreException {</span>
<span class="fc" id="L125">			this.transformer = transformer;</span>
		
<span class="fc" id="L127">		}</span>

		@Override
		public void visitHeader(SamVisitorCallback callback, SamHeader header) {
			
<span class="fc" id="L132">			Collection&lt;SamReferenceSequence&gt; referenceSequences = header.getReferenceSequences();</span>
<span class="fc" id="L133">			int capacity = MapUtil.computeMinHashMapSizeWithoutRehashing(referenceSequences.size());</span>
<span class="fc" id="L134">			referenceNames = new HashSet&lt;&gt;(capacity);</span>
<span class="fc" id="L135">			paddedReferenceSequences = new HashMap&lt;&gt;(capacity);</span>
<span class="fc" id="L136">			gapOffsetMap = new HashMap&lt;&gt;(capacity);</span>
<span class="fc bfc" id="L137" title="All 2 branches covered.">			for(SamReferenceSequence refSeq : referenceSequences){</span>
<span class="fc" id="L138">				referenceNames.add(refSeq.getName());</span>
<span class="fc" id="L139">			}</span>
			
<span class="fc" id="L141">		}</span>
			
		

	

		private boolean isReference(SamRecord record){
			//according to the SAMv1 spec
			//padded references should have
			//an identical RNAME, POS set to 1 and FLAG to 516 (filtered and unmapped)
<span class="pc bpc" id="L151" title="1 of 2 branches missed.">			return referenceNames.contains(record.getQueryName()) </span>
<span class="pc bpc" id="L152" title="1 of 2 branches missed.">					&amp;&amp; record.getStartPosition() == 1</span>
<span class="pc bpc" id="L153" title="1 of 2 branches missed.">					&amp;&amp; record.getFlags().maybeReferenceSequence();</span>
				
		}
		@Override
		public void visitRecord(SamVisitorCallback callback, SamRecord record, VirtualFileOffset start,
				VirtualFileOffset end) {
			
<span class="fc bfc" id="L160" title="All 2 branches covered.">			if(record.isPrimary()){</span>
				
<span class="fc bfc" id="L162" title="All 2 branches covered.">				if(record.mapped()){</span>
<span class="fc" id="L163">					String refName = record.getReferenceName();</span>
					
<span class="fc" id="L165">					NucleotideSequence referenceSeq = paddedReferenceSequences.get(refName);</span>
<span class="pc bpc" id="L166" title="1 of 2 branches missed.">					if(referenceSeq ==null){</span>
<span class="nc" id="L167">						throw new IllegalStateException(&quot;error padded reference sequence not defined in sam/bam file : &quot; + refName);</span>
					}
<span class="fc" id="L169">					Direction dir = record.getDirection();</span>
<span class="fc" id="L170">					Cigar cigar = record.getCigar();</span>
<span class="fc" id="L171">					int rawLength = cigar.getUnpaddedReadLength(ClipType.RAW);</span>
					Range validRange;
					//padded sams use gapped offset so don't have to convert
<span class="fc" id="L174">					int gappedStartOffset = record.getStartPosition()-1;</span>
					
					
					NucleotideSequence rawSequence;
					QualitySequence quals;
<span class="fc bfc" id="L179" title="All 2 branches covered.">					if(dir == Direction.FORWARD){</span>
<span class="fc" id="L180">						rawSequence = record.getSequence();</span>
<span class="fc" id="L181">						quals = record.getQualities();</span>
<span class="fc" id="L182">						validRange = cigar.getValidRange();</span>
					}else{
<span class="fc" id="L184">						rawSequence = new NucleotideSequenceBuilder(record.getSequence())</span>
<span class="fc" id="L185">											.reverseComplement()</span>
<span class="fc" id="L186">											.build();</span>
<span class="pc bpc" id="L187" title="1 of 2 branches missed.">						if(record.getQualities() ==null){</span>
<span class="fc" id="L188">							quals = null;</span>
						}else{
<span class="nc" id="L190">							quals = new QualitySequenceBuilder(record.getQualities())</span>
<span class="nc" id="L191">											.reverse()</span>
<span class="nc" id="L192">											.build();</span>
						}
<span class="fc" id="L194">						validRange = AssemblyUtil.reverseComplementValidRange(cigar.getValidRange(), rawLength);</span>
					}
				//	System.out.println(record.getQueryName());
<span class="fc" id="L197">					NucleotideSequence gappedReadSequence =removeHardClipsFrom(cigar).toGappedTrimmedSequence(record.getSequence());</span>
					
					//if the read is mated SAM doesn't put the /1 or /2 ?
					//what happens in CASAVA 1.8 reads?
<span class="fc" id="L201">					String readName = record.getQueryName();</span>
<span class="fc" id="L202">					SamRecordFlags flags = record.getFlags();</span>
					//from the SAMv1 spec
					//The leftmost segment has a plus sign and the rightmost has a
					//minus sign. The sign of segments in the middle is undefined. 
					//It is set as 0 for single-segment
					//template or when the information is unavailable.
<span class="fc bfc" id="L208" title="All 2 branches covered.">					if(flags.contains(SamRecordFlag.HAS_MATE_PAIR)){</span>
<span class="fc bfc" id="L209" title="All 2 branches covered.">						if(record.getObservedTemplateLength() &gt;=0){</span>
							//first read
<span class="fc" id="L211">							readName +=&quot;/1&quot;;</span>
						}else{
<span class="fc" id="L213">							readName +=&quot;/2&quot;;</span>
						}
						
					}
					//update valid range?
<span class="fc" id="L218">					transformer.aligned(readName, rawSequence, quals, null, null, </span>
							refName, 
							gappedStartOffset, 
							dir, 
							gappedReadSequence, 
							
							new ReadInfo(validRange, rawLength));
					
<span class="pc bpc" id="L226" title="1 of 2 branches missed.">				}else if(isReference(record)){</span>
					//add padded sequence to our map of references
<span class="fc" id="L228">					String name = record.getQueryName();</span>
<span class="fc" id="L229">					Cigar cigar = record.getCigar();</span>
<span class="fc" id="L230">					NucleotideSequence gappedSequence = cigar.toGappedTrimmedSequence(record.getSequence());</span>
<span class="fc" id="L231">					paddedReferenceSequences.put(name, gappedSequence);</span>
<span class="fc" id="L232">					gapOffsetMap.put(name, new GrowableIntArray(gappedSequence.getGapOffsets()));</span>
					
<span class="fc" id="L234">					transformer.referenceOrConsensus(name, gappedSequence);</span>
<span class="fc" id="L235">				}else{</span>
<span class="nc" id="L236">					transformer.notAligned(record.getQueryName(), record.getSequence(), record.getQualities(), null, null);</span>
					
				}
			}
				
<span class="fc" id="L241">		}</span>

		private Cigar removeHardClipsFrom(Cigar cigar) {
<span class="fc" id="L244">			return new Cigar.Builder(cigar)</span>
<span class="fc" id="L245">					.removeHardClips()</span>
<span class="fc" id="L246">					.build();</span>
		}

		@Override
		public void visitEnd() {
<span class="fc" id="L251">			transformer.endAssembly();			</span>
<span class="fc" id="L252">		}</span>
		
		@Override
		public void halted() {
<span class="nc" id="L256">			transformer.endAssembly();		</span>
<span class="nc" id="L257">		}</span>

	
	
}
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.5.201505241946</span></div></body></html>