<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../.resources/report.css" type="text/css"/><link rel="shortcut icon" href="../.resources/report.gif" type="image/gif"/><title>TraceInfoField.java</title><link rel="stylesheet" href="../.resources/prettify.css" type="text/css"/><script type="text/javascript" src="../.resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../.sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">Jillion</a> &gt; <a href="index.source.html" class="el_package">org.jcvi.jillion.experimental.trace.archive2</a> &gt; <span class="el_source">TraceInfoField.java</span></div><h1>TraceInfoField.java</h1><pre class="source lang-java linenums">/*******************************************************************************
 * Jillion development code
 * 
 * This code may be freely distributed and modified under the
 * terms of the GNU Lesser General Public License.  This should
 * be distributed with the code.  If you do not have a copy,
 *  see:
 * 
 *          http://www.gnu.org/copyleft/lesser.html
 * 
 * 
 * Copyright for this code is held jointly by the individual authors.  These should be listed in the @author doc comments.
 * 
 * Information about Jillion can be found on its homepage
 * 
 *         http://jillion.sourceforge.net
 * 
 * Contributors:
 *     Danny Katzel - initial API and implementation
 ******************************************************************************/
/*
 * Created on Sep 3, 2009
 *
 * @author dkatzel
 */
package org.jcvi.jillion.experimental.trace.archive2;

import java.util.Locale;

/**
 * All currently supported Fields in a Trace Info XML
 * document.
 * @author dkatzel
 * @see &lt;a href=&quot;http://www.ncbi.nlm.nih.gov/Traces/trace.cgi?cmd=show&amp;f=rfc&amp;m=doc&amp;s=rfc&quot;&gt;
 Trace Archive RFC Documentation&lt;/a&gt;
 *
 */
<span class="fc" id="L38">public enum TraceInfoField {</span>
    /**
     * Genbank/EMBL/DDBJ accession number
     */
<span class="fc" id="L42">     ACCESSION,</span>
     /**
     * The forward amplification primer sequence
     */
<span class="fc" id="L46">     AMPLIFICATION_FORWARD,</span>
     /**
     * The reverse amplification primer sequence
     */
<span class="fc" id="L50">     AMPLIFICATION_REVERSE,</span>
     /**
     * The expected amplification size for a pair of 
                         primers
     */
<span class="fc" id="L55">     AMPLIFICATION_SIZE,</span>
     /**
     * Anonymous ID for an individual.
     */
<span class="fc" id="L59">     ANONYMIZED_ID,</span>
     /**
     * Number of times the 
                 sequencing project has been attempted by the center and/or submitted to the 
                 Trace Archive.
     */
<span class="fc" id="L65">     ATTEMPT,</span>
     /**
     * File name with base calls
     */
<span class="fc" id="L69">     BASE_FILE,</span>
     /**
     * Name of the sequencing center.
     */
<span class="fc" id="L73">     CENTER_NAME,</span>
     /**
     * Center defined project name
     */
<span class="fc" id="L77">     CENTER_PROJECT,</span>
     /**
     * Description of the chemistry used in the sequencing reaction
     */
<span class="fc" id="L81">     CHEMISTRY,</span>
     /**
     * Type of chemistry used in the sequencing reaction
     */
<span class="fc" id="L85">     CHEMISTRY_TYPE,</span>
     /**
     * Chromosome to which the 
                 trace is assigned
     */
<span class="fc" id="L90">     CHROMOSOME,</span>
     /**
     * Left clip of the read, in 
                 base pairs, based on quality analysis
     */
<span class="fc" id="L95">     CLIP_QUALITY_LEFT,</span>
     /**
     * Right clip of the read, 
                 in base pairs, based on quality analysis.
     */
<span class="fc" id="L100">     CLIP_QUALITY_RIGHT,</span>
     /**
     * Left clip of the read, in 
                 base pairs, based on vector sequence.
     */
<span class="fc" id="L105">     CLIP_VECTOR_LEFT,</span>
     /**
     * Right clip of the read, 
                 in base pairs, based on vector sequence.
     */
<span class="fc" id="L110">     CLIP_VECTOR_RIGHT,</span>
     /**
     * The name of the clone 
                 from which the trace was derived
     */
<span class="fc" id="L115">     CLONE_ID,</span>
     /**
     * Semi-colon delimited 
                 list of clones if the Strategy is PoolClone.
     */
<span class="fc" id="L120">     CLONE_ID_LIST,</span>
     /**
     * The full date, in &quot;Mar  2 2006 12:00AM&quot; format, 
                 on which an environmental sample was collected.
     */
<span class="fc" id="L125">     COLLECTION_DATE,</span>
     /**
     * Repository ( 
                 GenBank/EMBL/DDBJ) accession identifier for the cloning vector.
     */
<span class="fc" id="L130">     CVECTOR_ACCESSION,</span>
     /**
     * Center defined code 
                 for the cloning vector
     */
<span class="fc" id="L135">     CVECTOR_CODE,</span>
     /**
     * Depth (in meters) at which an 
                 environmental sample was collected.
     */
<span class="fc" id="L140">     DEPTH,</span>
     /**
     * Elevation (in meters) at which 
                 an environmental sample was collected
     */
<span class="fc" id="L145">     ELEVATION,</span>
     /**
     * Type of 
                 environment from which an environmental sample was collected.
     */
<span class="fc" id="L150">     ENVIRONMENT_TYPE,</span>
     /**
     * Extra ancillary information wrapped 
                            around in a EXTENDED_DATA block, where actual values are provided 
                            with a special &amp;lt;field&amp;gt; tag
     */
<span class="fc" id="L156">     EXTENDED_DATA,</span>
     /**
     * File describing 
                 the features and their locations on a chip.
     */
<span class="fc" id="L161">     FEATURE_ID_FILE,</span>
     /**
     * Reference to a common 
                          FEATURE_ID_FILE which
                          should be submitted first.
     */
<span class="fc" id="L167">     FEATURE_ID_FILE_NAME,</span>
     /**
     * File giving 
                 the signal and variance for features on a chip.
     */
<span class="fc" id="L172">     FEATURE_SIGNAL_FILE,</span>
     /**
     * Reference to a common 
                          FEATURE_SIGNAL_FILE which
                          should be submitted first.
     */
<span class="fc" id="L178">     FEATURE_SIGNAL_FILE_NAME,</span>
     /**
     * Gene name or some other common identifier
     */
<span class="fc" id="L182">     GENE_NAME,</span>
     /**
     * The largest filter 
                 used to stratify an environmental sample.
     */
<span class="fc" id="L187">     HI_FILTER_SIZE,</span>
     /**
     * The condition of 
                 the host from which an environmental sample was obtained.
     */
<span class="fc" id="L192">     HOST_CONDITION,</span>
     /**
     * Unique identifier for the 
                 specific host from which an environmental sample was taken.
     */
<span class="fc" id="L197">     HOST_ID,</span>
     /**
     * Specific location 
                 on the host from which an environmental sample was collected.
     */
<span class="fc" id="L202">     HOST_LOCATION,</span>
     /**
     * The host from which an environmental sample was obtained.
     */
<span class="fc" id="L206">     HOST_SPECIES,</span>
     /**
     * Publicly available identifier to denote a specific
                    individual or sample from which a trace was derived.
     */
<span class="fc" id="L211">     INDIVIDUAL_ID,</span>
     /**
     * Flanking sequence at the cloning junction.
     */
<span class="fc" id="L215">     INSERT_FLANK_LEFT,</span>
     /**
     * Flanking sequence at the cloning junction.
     */
<span class="fc" id="L219">     INSERT_FLANK_RIGHT,</span>
     /**
     * Expected size of the insert (referred to by the value in the TEMPLATE_ID field) in base pairs.
     */
<span class="fc" id="L223">     INSERT_SIZE,</span>
     /**
     * Approximate standard deviation of value in INSERT_SIZE field
     */
<span class="fc" id="L227">     INSERT_STDEV,</span>
     /**
     * The latitude measurement (using 
                 standard GPS notation) from which a sample was collected.
     */
<span class="fc" id="L232">     LATITUDE,</span>
     /**
     * The source of the clone identified in the CLONE_ID field
     */
<span class="fc" id="L236">     LIBRARY_ID,</span>
     /**
     * The longitude measurement (using standard GPS notation) from which a sample
                     was collected.
     */
<span class="fc" id="L241">     LONGITUDE,</span>
     /**
     * The smallest filter size used to stratify an environmental sample.
     */
<span class="fc" id="L245">     LO_FILTER_SIZE,</span>
     /**
     * Project ID generated by the  
                                  Genome Project database at NCBI/NLM/NIH
     */
<span class="fc" id="L250">     NCBI_PROJECT_ID,</span>
     /**
     * Description of species for BARCODE project 
                                    from which trace is derived.
     */
<span class="fc" id="L255">     ORGANISM_NAME,</span>
     /**
     * Name of file that contains the list of peak values.
     */
<span class="fc" id="L259">     PEAK_FILE,</span>
     /**
     * The pH at which an environmental sample was collected.
     */
<span class="fc" id="L263">     PH,</span>
     /**
     * Id to group traces picked at the same time.
     */
<span class="fc" id="L267">     PICK_GROUP_ID,</span>
     /**
     * Country in which the  biological sample was collected and/or common name for a given location.
     */
<span class="fc" id="L271">     PLACE_NAME,</span>
     /**
     * Submitter defined plate id
     */
<span class="fc" id="L275">     PLATE_ID,</span>
     /**
     * Center provided id 
                 to designate a population from which a trace (or group of traces) was derived.
     */
<span class="fc" id="L280">     POPULATION_ID,</span>
     /**
     * ID that defines 
                 groups of traces prepared at the same time.
     */
<span class="fc" id="L285">     PREP_GROUP_ID,</span>
     /**
     * The primer sequence (used 
                 in the sequencing reaction)
     */
<span class="fc" id="L290">     PRIMER,</span>
     /**
     * Identifier for the sequencing primer used.
     */
<span class="fc" id="L294">     PRIMER_CODE,</span>
     /**
     * A ';' delimited list 
                 of primers used in a mapping experiment (such as AFLP)
     */
<span class="fc" id="L299">     PRIMER_LIST,</span>
     /**
     * The program used to create the trace file.
     */
<span class="fc" id="L303">     PROGRAM_ID,</span>
     /**
     * Term by which to group traces from different centers
                  based on a common project.
     */
<span class="fc" id="L308">     PROJECT_NAME,</span>
     /**
     * Name of file containing the quality scores.
     */
<span class="fc" id="L312">     QUAL_FILE,</span>
     /**
     * Reference 
                 accession (use accession and version to specify a particular instance of a 
                 sequence) used as the basis for a re-sequencing project. In case of Comparative strategy show the basis for primers design.
     */
<span class="fc" id="L318">     REFERENCE_ACCESSION,</span>
     /**
     * Finish position for a particular amplicon in
                      re-sequencing or comparative projects.
     */
<span class="fc" id="L323">     REFERENCE_ACC_MAX,</span>
     /**
     * Start position for a particular amplicon in
                         re-sequencing or comparative projects.
     */
<span class="fc" id="L328">     REFERENCE_ACC_MIN,</span>
     /**
     * Sequence offset of accession specified in
                     REFERENCE_ACCESSION field to define the coordinate start position used 
                              as the basis for a re-sequencing project.
     */
<span class="fc" id="L334">     REFERENCE_OFFSET,</span>
     /**
     * Finish position for a entire
                    re-sequencing region. This region may include several amplicons.
     */
<span class="fc" id="L339">     REFERENCE_SET_MAX,</span>
     /**
     * Start position for a entire
      re-sequencing region. This region may include several amplicons.
     */
<span class="fc" id="L344">     REFERENCE_SET_MIN,</span>
     /**
     * Date the sequencing reaction was run.
     */
<span class="fc" id="L348">     RUN_DATE,</span>
     /**
     * ID used to group traces run on the same machine.
     */
<span class="fc" id="L352">     RUN_GROUP_ID,</span>
     /**
     * Lane or capillary of the trace
     */
<span class="fc" id="L356">     RUN_LANE,</span>
     /**
     * ID of the specific sequencing machine on which a trace was obtained
     */
<span class="fc" id="L360">     RUN_MACHINE_ID,</span>
     /**
     * Type or model of machine on which a trace was obtained.
     */
<span class="fc" id="L364">     RUN_MACHINE_TYPE,</span>
     /**
     * The salinity at which an environmental sample was collected 
                                     measured in parts per thousand units (promille).
     */
<span class="fc" id="L369">     SALINITY,</span>
     /**
     * Center specified M13/PUC library that is actually sequenced.
     */
<span class="fc" id="L373">     SEQ_LIB_ID,</span>
     /**
     * Source of the DNA
     */
<span class="fc" id="L377">     SOURCE_TYPE,</span>
     /**
     * Description of species from which trace is derived
     */
<span class="fc" id="L381">     SPECIES_CODE,</span>
     /**
     * Strain from which a trace is derived.
     */
<span class="fc" id="L385">     STRAIN,</span>
     /**
     * Experimental STRATEGY
     */
<span class="fc" id="L389">     STRATEGY,</span>
     /**
     * Type of submission
     */
<span class="fc" id="L393">     SUBMISSION_TYPE,</span>
     /**
     * GenBank/EMBL/DDBJ accession of the sequencing vector.
     */
<span class="fc" id="L397">     SVECTOR_ACCESSION,</span>
     /**
     * Center defined code for the sequencing vector
     */
<span class="fc" id="L401">     SVECTOR_CODE,</span>
     /**
     * The temperature (in &lt;sup&gt;o&lt;/sup&gt;C) at which an environmental sample was collected.
     */
<span class="fc" id="L405">     TEMPERATURE,</span>
     /**
     * Submitter defined identifier for the sequencing template.
     */
<span class="fc" id="L409">     TEMPLATE_ID,</span>
     /**
     * Defines the end of the 
                 template contained in the read.
     */
<span class="fc" id="L414">     TRACE_END,</span>
     /**
     * Filename with the 
                 trace, relative to the top of the volume.
     */
<span class="fc" id="L419">     TRACE_FILE,</span>
     /**
     * Format of the trace file.
     */
<span class="fc" id="L423">     TRACE_FORMAT,</span>
     /**
     * Center defined trace identifier.
     */
<span class="fc" id="L427">     TRACE_NAME,</span>
     /**
     * Sequencing strategy by which 
                                     the trace was obtained.
     */
<span class="fc" id="L432">     TRACE_TYPE_CODE,</span>
     /**
     * GenBank/EMBL/DDBJ 
                 accession for transposon used in generating sequencing template.
     */
<span class="fc" id="L437">     TRANSPOSON_ACC,</span>
     /**
     * Center defined 
                 code for transposon used in generating sequencing template.
     */
<span class="fc" id="L442">     TRANSPOSON_CODE,</span>
     /**
     * Center defined well 
                 identifier for the sequencing reaction.
     */
<span class="fc" id="L447">     WELL_ID,</span>
     /**
     * Length of the trace in base pairs.
     */
<span class="fc" id="L451">     BASECALL_LENGTH,</span>
     /**
     * Number of  base pairs for which quality score exceed 20.
     */
<span class="fc" id="L455">     BASES_20,</span>
     /**
     * Number of  base pairs for which quality score exceed 40.
     */
<span class="fc" id="L459">     BASES_40,</span>
     /**
     * Number of  base pairs for which quality score exceed 60.
     */
<span class="fc" id="L463">     BASES_60,</span>
     /**
     * Date on which the data was loaded.
     */
<span class="fc" id="L467">     LOAD_DATE,</span>
     /**
     * TI's of the reads obtained from the other end of the same template.
     */
<span class="fc" id="L471">     MATE_PAIR,</span>
     
<span class="fc" id="L473">     MATE_TI,</span>
     
<span class="fc" id="L475">     MATE_NAME,</span>
     
<span class="fc" id="L477">     MATE_TEMPLATE_ID,</span>
     
<span class="fc" id="L479">     MATE_TRACE_END,</span>
     /**
     * TI that replaced the 
                 current TI as &quot;active&quot;
     */
<span class="fc" id="L484">     REPLACED_BY,</span>
     /**
     * indicates the status of the trace
     */
<span class="fc" id="L488">     STATE,</span>
     /**
     * NCBI Taxonomy ID.
     */
<span class="fc" id="L492">     TAXID,</span>
     /**
     * Trace unique internal Identifier.
     */
<span class="fc" id="L496">     TI,</span>
     /**
     * Date on which the data was updated/replaced.
     */
<span class="fc" id="L500">     UPDATE_DATE,</span>
     /**
     * Public identifier for a given version of a genome assembly
     */
<span class="fc" id="L504">     ASSEMBLY_ID,</span>
     /**
     * 
     */
<span class="fc" id="L508">     CHROMOSOME_REGION,</span>
     /**
      * Name of subspecies.
      */
<span class="fc" id="L512">     SUBSPECIES_ID,</span>
     
     /**
     * Direction of the read.
     */
<span class="fc" id="L517">     TRACE_DIRECTION;</span>
     
     public static TraceInfoField parseTraceInfoField(String traceInfoField){
<span class="fc" id="L520">         return TraceInfoField.valueOf(traceInfoField.toUpperCase(Locale.US));</span>
     }

    @Override
    public String toString() {
<span class="fc" id="L525">        return this.name().toLowerCase(Locale.US);</span>
    }

}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.5.201505241946</span></div></body></html>