<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../.resources/report.css" type="text/css"/><link rel="shortcut icon" href="../.resources/report.gif" type="image/gif"/><title>NucleotideFastaFileDataStore.java</title><link rel="stylesheet" href="../.resources/prettify.css" type="text/css"/><script type="text/javascript" src="../.resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../.sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">Jillion</a> &gt; <a href="index.source.html" class="el_package">org.jcvi.jillion.fasta.nt</a> &gt; <span class="el_source">NucleotideFastaFileDataStore.java</span></div><h1>NucleotideFastaFileDataStore.java</h1><pre class="source lang-java linenums">/*******************************************************************************
 * Jillion development code
 * 
 * This code may be freely distributed and modified under the
 * terms of the GNU Lesser General Public Licence.  This should
 * be distributed with the code.  If you do not have a copy,
 *  see:
 * 
 *          http://www.gnu.org/copyleft/lesser.html
 * 
 * 
 * Copyright for this code is held jointly by the individual authors.  These should be listed in the @author doc comments.
 * 
 * Information about Jillion can be found on its homepage
 * 
 *         http://jillion.sourceforge.net
 * 
 * Contributors:
 *     Danny Katzel - initial API and implementation
 ******************************************************************************/
package org.jcvi.jillion.fasta.nt;

import java.io.File;
import java.io.IOException;
import java.util.Optional;

import org.jcvi.jillion.core.datastore.DataStoreProviderHint;
/**
 * A {@link NucleotideFastaDataStore} where all the {@link NucleotideFastaRecord}s
 * in this datastore belong to the same fasta file.
 * @author dkatzel
 * @since 5.3
 */
public interface NucleotideFastaFileDataStore extends NucleotideFastaDataStore{
    
    /**
     * Create a {@link NucleotideFastaDataStore} of all the records
     * in the given fasta file.  Warning! This usually stores all the records in memory
     * use {@link #fromFile(File, DataStoreProviderHint)} or {@link NucleotideFastaFileDataStoreBuilder}
     * to handle the file or datastore implementation differently.
     * 
     * @param fastaFile the fasta file make a {@link NucleotideFastaDataStore} with;
     * can not be null and must exist.
     * 
     * @apiNote As of Jillion 5.1, this will also look for a FastaIndex ({@code .fai})
     * file using the standard naming conventions and is equivalent to
     * &lt;pre&gt;
     * new NucleotideFastaFileDataStoreBuilder(fastaFile, new File(fastaFile.getParentFile(), fastaFile.getName() + &quot;.fai&quot;));
     * &lt;/pre&gt;
     * 
     * @throws IOException if the fasta file does not exist, or can not be read.
     * @throws NullPointerException if fastaFile is null.
     * @return a new NucleotideFastaDataStore; will never be null.
     * @since 5.3
     * 
     * @see NucleotideFastaFileDataStoreBuilder
     * @see #fromFile(File, DataStoreProviderHint)
     */
    public static NucleotideFastaFileDataStore fromFile(File fastaFile) throws IOException{
<span class="fc" id="L60">        return new NucleotideFastaFileDataStoreBuilder(fastaFile).build();</span>
    }
    /**
     * Create a {@link NucleotideFastaDataStore} of all the records
     * in the given fasta file with the given {@link DataStoreProviderHint}
     * to help jillion choose a datastore implementation.  If filtering records
     * is desired, pleas use {@link NucleotideFastaFileDataStoreBuilder}.
     * 
     * @param fastaFile the fasta file make a {@link NucleotideFastaDataStore} with;
     * can not be null and must exist.
     * 
     * @param hint the {@link DataStoreProviderHint} to use; can not be null.
     * 
     * @apiNote As of Jillion 5.1, this will also look for a FastaIndex ({@code .fai})
     * file using the standard naming conventions and is equivalent to
     * &lt;pre&gt;
     * new NucleotideFastaFileDataStoreBuilder(fastaFile, new File(fastaFile.getParentFile(), fastaFile.getName() + &quot;.fai&quot;));
     * &lt;/pre&gt;
     * 
     * @throws IOException if the fasta file does not exist, or can not be read.
     * @throws NullPointerException if any parameter is null.
     * @return a new NucleotideFastaDataStore; will never be null.
     * @since 5.3
     * 
     * @see NucleotideFastaFileDataStoreBuilder
     */
    public static NucleotideFastaFileDataStore fromFile(File fastaFile, DataStoreProviderHint hint) throws IOException{
<span class="nc" id="L87">        NucleotideFastaFileDataStoreBuilder builder =  new NucleotideFastaFileDataStoreBuilder(fastaFile);</span>
<span class="nc" id="L88">        builder.hint(hint);</span>
        
<span class="nc" id="L90">        return builder.build();</span>
    }
    
    /**
     * Get the actual {@link java.io.File} that this datastore
     * wraps.  Note, because of datastore filtering and other decorators, the file may contain additional reads
     * not present in this datastore and the {@link NucleotideFastaRecord}s that are present,
     *  may not match 100% to the input file.  It is also possible that this file
     *  uses some kind of compression that this datastore knows how to decompress.
     * 
     * @return an {@link Optional}  {@link java.io.File}, if it is known.
     * 
     * @since 5.3
     */
    Optional&lt;File&gt; getFile();
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.5.201505241946</span></div></body></html>