<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../.resources/report.css" type="text/css"/><link rel="shortcut icon" href="../.resources/report.gif" type="image/gif"/><title>AceFileUtil.java</title><link rel="stylesheet" href="../.resources/prettify.css" type="text/css"/><script type="text/javascript" src="../.resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../.sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">Jillion</a> &gt; <a href="index.source.html" class="el_package">org.jcvi.jillion.assembly.consed.ace</a> &gt; <span class="el_source">AceFileUtil.java</span></div><h1>AceFileUtil.java</h1><pre class="source lang-java linenums">/*******************************************************************************
 * Jillion development code
 * 
 * This code may be freely distributed and modified under the
 * terms of the GNU Lesser General Public License.  This should
 * be distributed with the code.  If you do not have a copy,
 *  see:
 * 
 *          http://www.gnu.org/copyleft/lesser.html
 * 
 * 
 * Copyright for this code is held jointly by the individual authors.  These should be listed in the @author doc comments.
 * 
 * Information about Jillion can be found on its homepage
 * 
 *         http://jillion.sourceforge.net
 * 
 * Contributors:
 *     Danny Katzel - initial API and implementation
 ******************************************************************************/
package org.jcvi.jillion.assembly.consed.ace;

import java.io.IOException;
import java.io.OutputStream;
import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;

import org.jcvi.jillion.assembly.AssembledRead;
import org.jcvi.jillion.assembly.AssemblyUtil;
import org.jcvi.jillion.assembly.Contig;
import org.jcvi.jillion.assembly.consed.phd.Phd;
import org.jcvi.jillion.assembly.consed.phd.PhdDataStore;
import org.jcvi.jillion.core.Direction;
import org.jcvi.jillion.core.Range;
import org.jcvi.jillion.core.Range.CoordinateSystem;
import org.jcvi.jillion.core.datastore.DataStoreException;
import org.jcvi.jillion.core.io.IOUtil;
import org.jcvi.jillion.core.qual.PhredQuality;
import org.jcvi.jillion.core.qual.QualitySequence;
import org.jcvi.jillion.core.qual.QualitySequenceBuilder;
import org.jcvi.jillion.core.qual.QualitySequenceDataStore;
import org.jcvi.jillion.core.residue.nt.Nucleotide;
import org.jcvi.jillion.core.residue.nt.NucleotideSequence;
import org.jcvi.jillion.core.util.iter.StreamingIterator;
import org.jcvi.jillion.internal.core.util.JillionUtil;

/**
 * {@code AceFileUtil} is a utility class to perform 
 * common operations on ace related objects.
 * @author dkatzel
 *
 *
 */
public final class AceFileUtil {
	
    /**
     * This is the default value in consed that is used to distinguish
     * between high and low quality basecalls.  In consed
     * high quality bases are represented by uppercase
     * letters and low quality by lowercase letters.
     * Currently that value is set to 26.
     */
<span class="fc" id="L70">    public static final PhredQuality ACE_DEFAULT_HIGH_QUALITY_THRESHOLD = PhredQuality.valueOf(26);</span>
    /**
     * The date format used in consed to represent chromatogram time stamps 
     * in phd records as well as the DS lines in an ace file.
     * A read's timestamps must be identical strings in both the phd
     * and the ace for consed to make the read editable and to see the qualities
     * in the align window.
     * &lt;p&gt;
     * Use {@link ThreadLocal} since each DateFormat instance
	 * is mutable and not Thread safe.
	 * This should let us avoid synchronization.
	 * &lt;/p&gt;
     */   
<span class="fc" id="L83">	private static ThreadLocal&lt;DateFormat&gt; CHROMAT_DATE_TIME_FORMATTER = new ThreadLocal&lt;DateFormat&gt;(){</span>

		  @Override
		  public DateFormat get() {
<span class="fc" id="L87">		   return super.get();</span>
		  }

		  @Override
		  protected DateFormat initialValue() {
<span class="fc" id="L92">			DateFormat format = new SimpleDateFormat(&quot;EEE MMM d HH:mm:ss yyyy&quot;,</span>
					Locale.US);

<span class="fc" id="L95">			format.setLenient(false);</span>
<span class="fc" id="L96">			return format;</span>
		}
		 };
	
    
    /**
	 * This is the timestamp format used in some consed 
     * tags.
     * Use {@link ThreadLocal} since each DateFormat instance
	 * is mutable and not Thread safe.
	 * This should let us avoid synchronization.
	 */
<span class="fc" id="L108">	private static ThreadLocal&lt;KnownFormats&gt; TAG_DATE_TIME_FORMATTER = new ThreadLocal&lt;KnownFormats&gt;(){</span>

		  @Override
		  public KnownFormats get() {
<span class="fc" id="L112">		   return super.get();</span>
		  }

		  @Override
		  protected KnownFormats initialValue() {

<span class="fc" id="L118">			  return new KnownFormats(&quot;yyMMdd:HHmmss&quot;, </span>
					  					&quot;yyMMdd&quot;);
		  }


		 };
	/**
	 * Wrapper around multiple {@link SimpleDateFormat}s that will
	 * try each one until it passes {@link #parse(String)}.
	 * 
	 * &lt;p&gt;
	 * Not thread-safe 
	 * &lt;/p&gt;
	 * @author dkatzel
	 *
	 */
    private static class KnownFormats{
    	private final List&lt;DateFormat&gt; knownFormats;
    	
<span class="fc" id="L137">    	public KnownFormats(String...formats){</span>
<span class="fc" id="L138">    		knownFormats = new ArrayList&lt;&gt;(formats.length);</span>
<span class="fc bfc" id="L139" title="All 2 branches covered.">    		for(String format : formats){</span>
<span class="fc" id="L140">    			DateFormat f = new SimpleDateFormat(format, Locale.US);</span>
<span class="fc" id="L141">    			f.setLenient(true);</span>
<span class="fc" id="L142">    			knownFormats.add(f);</span>
    		}
<span class="fc" id="L144">    	}</span>
    	
    	public Date parse(String dateString) throws ParseException{
<span class="pc bpc" id="L147" title="1 of 2 branches missed.">    		for(DateFormat f : knownFormats){</span>
    			try{
<span class="fc" id="L149">    				return f.parse(dateString);</span>
<span class="fc" id="L150">    			}catch(ParseException e){</span>
    				//keep going
    			}
<span class="fc" id="L153">    		}</span>
    		
<span class="nc" id="L155">    		throw new ParseException(&quot;Unparseable date: \&quot;&quot; + dateString + &quot;\&quot;&quot;, 0);</span>
    	}
    	
    	public String format(Date date){
    		//always use first one
<span class="fc" id="L160">    		return knownFormats.get(0).format(date);</span>
    	}
    }
		 
    private static final String CONTIG_HEADER = &quot;CO %s %d %d %d %s%n&quot;;
    /**
     * Initial size to use for StringBuilders used to construct
     * ace encoded read data.
     */
	private static final int INITIAL_READ_BUFFER_SIZE = 1024;
   
    
<span class="nc" id="L172">    private AceFileUtil(){</span>
		//private constructor.
<span class="nc" id="L174">	}</span>
    /**
     * Parse the date of a consed formatted phd date time stamp.
     * @param text the timestamp as text
     * @return the equivalent time as a {@link Date}.
     * @throws ParseException if there is a problem parsing the text.
     */
    public static Date parsePhdDate(String text) throws ParseException{
<span class="fc" id="L182">    	return CHROMAT_DATE_TIME_FORMATTER.get().parse(text);</span>
    }
    /**
     * format the given {@link Date} representing
     * a phd date time stamp into a consed formatted phd time stamp.
     * @param date the phd date to format.
     * @return the equivalent time as a String.
     */
    public static String formatPhdDate(Date date){
<span class="fc" id="L191">    	return CHROMAT_DATE_TIME_FORMATTER.get().format(date);</span>
    }
    
    /**
     * Parse the date of a consed formatted tag date time stamp.
     * @param text the timestamp as text
     * @return the equivalent time as a {@link Date}.
     * @throws ParseException if there is a problem parsing the text.
     */
    public static Date parseTagDate(String text) throws ParseException{
<span class="fc" id="L201">    	return TAG_DATE_TIME_FORMATTER.get().parse(text);</span>
    }
    /**
     * Format the given {@link Date} representing
     * a date time stamp for a consed tag
     * into a consed formatted time stamp.
     * @param date the date to format.
     * @return the equivalent time as a String.
     */
    public static String formatTagDate(Date date){
<span class="fc" id="L211">    	return TAG_DATE_TIME_FORMATTER.get().format(date);</span>
    }
    /**
     * Convert a {@link NucleotideSequence} into a string
     * where the gaps are represented by '*'s like ace files require.
     * @param basecalls input basecalls, can not be null.
     * @return a String with the all uppercase basecalls except all the '-'
     * have been replaced by '*'.
     * @throws NullPointerException if basecalls are null.
     */
    public static String convertToAcePaddedBasecalls(NucleotideSequence basecalls){
<span class="fc" id="L222">        return convertToAcePaddedBasecalls(basecalls,null);</span>
     }
    /**
     * Convert a {@link NucleotideSequence} into a string
     * where the gaps are represented by '*'s like ace files require.
     * If the optional qualities list is provided, then the returned 
     * String will return a basecalls in both upper and lowercase
     * depending on the quality value as determined by {@link AceFileUtil#ACE_DEFAULT_HIGH_QUALITY_THRESHOLD}.
     * 
     * @param basecalls input basecalls, can not be null or contain any null elements.
     * @param optionalQualities optional ungapped quality list in the same
     * orientation as the basecalls; or null if no qualities are to be used.
     * @return a String with the all uppercase basecalls except all the '-'
     * have been replaced by '*'.
     * @throws NullPointerException if basecalls are null or any Nucleotide in the basecall list 
     * is null or if the optionalQualities list is not null but contains a null in the list.
     * @throws IllegalArgumentException if optionalQualities is provided but does not have 
     * enough ungapped qualities to cover all the ungapped basecalls in the input list.
     */
     public static String convertToAcePaddedBasecalls(NucleotideSequence basecalls,QualitySequence optionalQualities){
<span class="fc" id="L242">         long length = basecalls.getLength();</span>
<span class="fc" id="L243">		StringBuilder result = new StringBuilder((int)length);</span>
<span class="fc bfc" id="L244" title="All 2 branches covered.">		Iterator&lt;PhredQuality&gt; qualityIterator = optionalQualities ==null?null : optionalQualities.iterator();</span>
<span class="fc bfc" id="L245" title="All 2 branches covered.">         for(Nucleotide base : basecalls){</span>
<span class="fc bfc" id="L246" title="All 2 branches covered.">             if(base == Nucleotide.Gap){</span>
<span class="fc" id="L247">                 result.append('*');</span>
             }
             else{
<span class="fc bfc" id="L250" title="All 2 branches covered.">                 if(optionalQualities==null){</span>
<span class="fc" id="L251">                	 result.append(base);</span>
                 }else{                 
<span class="pc bpc" id="L253" title="1 of 2 branches missed.">                     if(!qualityIterator.hasNext()){</span>
<span class="nc" id="L254">                         throw new IllegalArgumentException(</span>
<span class="nc" id="L255">                                 String.format(&quot;not enough ungapped qualities for input basecalls found only %d qualities&quot;,optionalQualities.getLength()));</span>
                     }
<span class="fc" id="L257">                     PhredQuality quality =qualityIterator.next();</span>
<span class="fc bfc" id="L258" title="All 2 branches covered.">                     if(quality.compareTo(ACE_DEFAULT_HIGH_QUALITY_THRESHOLD)&lt;0){</span>
<span class="fc" id="L259">                         result.append(base.toString().toLowerCase(Locale.ENGLISH));</span>
                     }
                     else{
<span class="fc" id="L262">                         result.append(base);</span>
                     }
                 }
             }
<span class="fc" id="L266">         }</span>
         
<span class="fc" id="L268">         String consedBasecalls= result.toString().replaceAll(&quot;(.{50})&quot;, &quot;$1&quot;+String.format(&quot;%n&quot;));</span>
<span class="fc bfc" id="L269" title="All 2 branches covered.">         if(length %50 ==0){</span>
             //if the last line is full, then we will have an extra %n
             //so strip it off
<span class="fc" id="L272">             return consedBasecalls.substring(0,consedBasecalls.length()-1);</span>
         }
<span class="fc" id="L274">         return consedBasecalls;</span>
     }

    
    private static String createPhdRecord(PhdInfo phdInfo){
<span class="fc" id="L279">        return String.format(&quot;DS CHROMAT_FILE: %s PHD_FILE: %s TIME: %s&quot;, </span>
                
<span class="fc" id="L281">                phdInfo.getTraceName(),</span>
<span class="fc" id="L282">                phdInfo.getPhdName(),</span>
<span class="fc" id="L283">                formatPhdDate(phdInfo.getPhdDate())</span>
                );
    }
    
    private static String createQualityRangeRecord(NucleotideSequence gappedValidBases, 
            Range ungappedValidRange, Direction dir, long ungappedFullLength){
<span class="fc" id="L289">        int numberOfGaps = gappedValidBases.getNumberOfGaps();</span>
<span class="fc" id="L290">        Range gappedValidRange =buildGappedValidRangeFor(</span>
                ungappedValidRange,numberOfGaps,dir,ungappedFullLength);

        
<span class="fc" id="L294">       return String.format(&quot;QA %d %d %d %d&quot;,</span>
<span class="fc" id="L295">                gappedValidRange.getBegin(CoordinateSystem.RESIDUE_BASED), gappedValidRange.getEnd(CoordinateSystem.RESIDUE_BASED),</span>
<span class="fc" id="L296">                gappedValidRange.getBegin(CoordinateSystem.RESIDUE_BASED), gappedValidRange.getEnd(CoordinateSystem.RESIDUE_BASED)</span>
                );
    }
    private static Range buildGappedValidRangeFor(Range ungappedValidRange, int numberOfGaps,Direction dir, long ungappedFullLength){
<span class="fc" id="L300">       Range gappedValidRange=  Range.of( </span>
<span class="fc" id="L301">               ungappedValidRange.getBegin(),</span>
<span class="fc" id="L302">               ungappedValidRange.getEnd()+numberOfGaps);</span>
        
<span class="fc bfc" id="L304" title="All 2 branches covered.">        if(dir==Direction.REVERSE){</span>
<span class="fc" id="L305">            gappedValidRange = AssemblyUtil.reverseComplementValidRange(gappedValidRange, ungappedFullLength+numberOfGaps);</span>
           
        }
<span class="fc" id="L308">        return gappedValidRange;</span>
    }
    public static synchronized String createAcePlacedReadRecord(String readId, AssembledRead placedRead, Phd phd, PhdInfo phdInfo){
        
<span class="pc bpc" id="L312" title="1 of 2 branches missed.">    	if(phd==null){</span>
<span class="nc" id="L313">    		throw new NullPointerException(&quot;phd can not be null for &quot; + readId);</span>
    	}
<span class="fc" id="L315">        NucleotideSequence nucleotideSequence = placedRead.getNucleotideSequence();</span>
<span class="fc" id="L316">		final NucleotideSequence gappedValidBasecalls = nucleotideSequence; </span>
<span class="fc" id="L317">        final Range ungappedValidRange = placedRead.getReadInfo().getValidRange();</span>
<span class="fc" id="L318">        final Direction dir = placedRead.getDirection(); </span>
<span class="fc" id="L319">        final NucleotideSequence fullBasecalls = phd.getNucleotideSequence();</span>
        
<span class="fc" id="L321">        final NucleotideSequence fullGappedValidRange = AssemblyUtil.buildGappedComplementedFullRangeBases(placedRead, fullBasecalls);</span>
        final QualitySequence qualities;
<span class="fc bfc" id="L323" title="All 2 branches covered.">        if(dir ==Direction.REVERSE){</span>
<span class="fc" id="L324">        	qualities = new QualitySequenceBuilder(phd.getQualitySequence())</span>
<span class="fc" id="L325">        						.reverse()</span>
<span class="fc" id="L326">        						.build();</span>
        }else{
<span class="fc" id="L328">        	qualities = phd.getQualitySequence();</span>
        }
        
<span class="fc" id="L331">        return new StringBuilder(INITIAL_READ_BUFFER_SIZE)</span>
<span class="fc" id="L332">        		.append(String.format(&quot;RD %s %d 0 0%n&quot;,</span>
                                            readId,
<span class="fc" id="L334">                                            fullGappedValidRange.getLength()))</span>
        
        
<span class="fc" id="L337">	        .append(String.format(&quot;%s%n%n&quot;,</span>
<span class="fc" id="L338">	                AceFileUtil.convertToAcePaddedBasecalls(fullGappedValidRange,qualities)))</span>
<span class="fc" id="L339">	        .append(String.format(&quot;%s%n&quot;,createQualityRangeRecord(</span>
	                gappedValidBasecalls,ungappedValidRange,dir, 
<span class="fc" id="L341">	                fullBasecalls.getUngappedLength())))</span>
<span class="fc" id="L342">	        .append(String.format(&quot;%s%n&quot;,createPhdRecord(phdInfo)))</span>
<span class="fc" id="L343">	        .toString();</span>
    }
    public static void writeAceContigHeader(String contigId, long consensusLength, long numberOfReads,
    		int numberOfBaseSegments, boolean isComplimented, OutputStream out) throws IOException{
<span class="fc" id="L347">    	writeString(String.format(CONTIG_HEADER, </span>
                contigId, 
<span class="fc" id="L349">                consensusLength,</span>
<span class="fc" id="L350">                numberOfReads,</span>
<span class="pc bpc" id="L351" title="1 of 2 branches missed.">                numberOfBaseSegments,</span>
                isComplimented? &quot;C&quot;:&quot;U&quot;),
                
                out);
<span class="fc" id="L355">    }</span>
    public static void writeAceFileHeader(long numberOfContigs, long numberOfReads, OutputStream out) throws IOException{
<span class="fc" id="L357">        writeString(String.format(&quot;AS %d %d%n%n&quot;, numberOfContigs, numberOfReads), out);</span>
<span class="fc" id="L358">    }</span>
    private static void writeString(String s, OutputStream out) throws IOException{
<span class="fc" id="L360">        out.write(s.getBytes(IOUtil.UTF_8));</span>
        
<span class="fc" id="L362">    }</span>
    
    
    public static void writeWholeAssemblyTag(
            WholeAssemblyAceTag wholeAssemblyTag, OutputStream out) throws IOException {
<span class="nc" id="L367">        writeString(String.format(&quot;WA{%n%s %s %s%n%s%n}%n&quot;, </span>
<span class="nc" id="L368">                wholeAssemblyTag.getType(),</span>
<span class="nc" id="L369">                wholeAssemblyTag.getCreator(),                </span>
<span class="nc" id="L370">                AceFileUtil.formatTagDate(wholeAssemblyTag.getCreationDate()),</span>
<span class="nc" id="L371">                wholeAssemblyTag.getData()), out);</span>

        
<span class="nc" id="L374">    }</span>
    public static void writeConsensusTag(ConsensusAceTag consensusTag,
            OutputStream out) throws IOException {
<span class="nc" id="L377">        StringBuilder tagBodyBuilder = new StringBuilder();</span>
<span class="nc bnc" id="L378" title="All 2 branches missed.">        if(consensusTag.getData() !=null){</span>
<span class="nc" id="L379">            tagBodyBuilder.append(consensusTag.getData());</span>
        }
<span class="nc bnc" id="L381" title="All 2 branches missed.">        if(!consensusTag.getComments().isEmpty()){</span>
<span class="nc bnc" id="L382" title="All 2 branches missed.">            for(String comment :consensusTag.getComments()){</span>
<span class="nc" id="L383">                tagBodyBuilder.append(String.format(&quot;COMMENT{%n%sC}%n&quot;,comment));            </span>
<span class="nc" id="L384">            }</span>
        }
<span class="nc" id="L386">        Range range = consensusTag.asRange();</span>
<span class="nc" id="L387">        writeString(String.format(&quot;CT{%n%s %s %s %d %d %s%s%n%s}%n&quot;, </span>
<span class="nc" id="L388">                consensusTag.getId(),</span>
<span class="nc" id="L389">                consensusTag.getType(),</span>
<span class="nc" id="L390">                consensusTag.getCreator(),</span>
<span class="nc" id="L391">                range.getBegin(),</span>
<span class="nc" id="L392">                range.getEnd(),</span>
<span class="nc" id="L393">                AceFileUtil.formatTagDate(consensusTag.getCreationDate()),</span>
<span class="nc bnc" id="L394" title="All 2 branches missed.">                consensusTag.isTransient()?&quot; NoTrans&quot;:&quot;&quot;,</span>
<span class="nc" id="L395">                        tagBodyBuilder.toString()), out);</span>
        
<span class="nc" id="L397">    }</span>
    public static void writeReadTag(ReadAceTag readTag, OutputStream out) throws IOException {
<span class="nc" id="L399">        Range range = readTag.asRange();</span>
<span class="nc" id="L400">    	writeString(String.format(&quot;RT{%n%s %s %s %d %d %s%n}%n&quot;, </span>
<span class="nc" id="L401">                        readTag.getId(),</span>
<span class="nc" id="L402">                        readTag.getType(),</span>
<span class="nc" id="L403">                        readTag.getCreator(),</span>
<span class="nc" id="L404">                        range.getBegin(),</span>
<span class="nc" id="L405">                        range.getEnd(),</span>
<span class="nc" id="L406">                        AceFileUtil.formatTagDate(readTag.getCreationDate())), out);</span>
        
<span class="nc" id="L408">    }</span>
    
    public static void writeAceContig(AceContig contig,
            PhdDataStore phdDataStore, 
            OutputStream out) throws IOException, DataStoreException{
<span class="fc" id="L413">        final NucleotideSequence consensus = contig.getConsensusSequence();</span>
<span class="fc" id="L414">        writeAceContigHeader(</span>
<span class="fc" id="L415">                contig.getId(), </span>
<span class="fc" id="L416">                consensus.getLength(),</span>
<span class="fc" id="L417">                contig.getNumberOfReads(),</span>
                0,
<span class="fc" id="L419">                contig.isComplemented(),</span>
                out);
<span class="fc" id="L421">        out.flush();</span>
<span class="fc" id="L422">        writeString(String.format(&quot;%s%n%n%n&quot;,AceFileUtil.convertToAcePaddedBasecalls(consensus)), out);</span>
<span class="fc" id="L423">        out.flush();</span>
<span class="fc" id="L424">        writeFakeUngappedConsensusQualities(consensus, out);</span>
<span class="fc" id="L425">        writeString(String.format(&quot;%n&quot;), out);</span>
<span class="fc" id="L426">        out.flush();</span>
<span class="fc" id="L427">        List&lt;IdAlignedReadInfo&gt; assembledFroms = getSortedAssembledFromsFor(contig);</span>
<span class="fc" id="L428">        StringBuilder assembledFromBuilder = new StringBuilder();</span>
<span class="fc" id="L429">        StringBuilder placedReadBuilder = new StringBuilder();</span>
        
<span class="fc bfc" id="L431" title="All 2 branches covered.">        for(IdAlignedReadInfo assembledFrom : assembledFroms){</span>
<span class="fc" id="L432">            String id = assembledFrom.getId();</span>
<span class="fc" id="L433">            final Phd phd = phdDataStore.get(id);</span>
<span class="fc" id="L434">            final AceAssembledRead realPlacedRead = contig.getRead(id);</span>
<span class="fc" id="L435">             long fullLength = realPlacedRead.getReadInfo().getUngappedFullLength();</span>
<span class="fc" id="L436">            assembledFromBuilder.append(createAssembledFromRecord(realPlacedRead,fullLength));</span>
<span class="fc" id="L437">            placedReadBuilder.append(createPlacedReadRecord(realPlacedRead, phd));</span>
<span class="fc" id="L438">        }</span>
<span class="fc" id="L439">        assembledFromBuilder.append(String.format(&quot;%n&quot;));</span>
<span class="fc" id="L440">        placedReadBuilder.append(String.format(&quot;%n&quot;));</span>
<span class="fc" id="L441">        writeString(assembledFromBuilder.toString(),out);</span>
<span class="fc" id="L442">        out.flush();</span>
<span class="fc" id="L443">        writeString(placedReadBuilder.toString(),out);</span>
<span class="fc" id="L444">        out.flush();</span>
<span class="fc" id="L445">    }</span>
    private static List&lt;IdAlignedReadInfo&gt; getSortedAssembledFromsFor(
            Contig&lt;AceAssembledRead&gt; contig){
<span class="fc" id="L448">        List&lt;IdAlignedReadInfo&gt; assembledFroms = new ArrayList&lt;IdAlignedReadInfo&gt;((int)contig.getNumberOfReads());</span>
<span class="fc" id="L449">        StreamingIterator&lt;AceAssembledRead&gt; iter = null;</span>
        try{
<span class="fc" id="L451">        	iter = contig.getReadIterator();</span>
<span class="fc bfc" id="L452" title="All 2 branches covered.">        	while(iter.hasNext()){</span>
<span class="fc" id="L453">        		AceAssembledRead read = iter.next();</span>
<span class="fc" id="L454">        		long fullLength =read.getReadInfo().getUngappedFullLength();</span>
<span class="fc" id="L455">	            assembledFroms.add(IdAlignedReadInfo.createFrom(read, fullLength));</span>
<span class="fc" id="L456">        	}</span>
        }finally{
<span class="pc" id="L458">        	IOUtil.closeAndIgnoreErrors(iter);</span>
<span class="fc" id="L459">        }</span>
<span class="fc" id="L460">        Collections.sort(assembledFroms);</span>
<span class="fc" id="L461">        return assembledFroms;</span>
    }
    
    private static void writeFakeUngappedConsensusQualities(NucleotideSequence consensus,
            OutputStream out) throws IOException {
<span class="fc" id="L466">        StringBuilder result = new StringBuilder((int)consensus.getLength());</span>
<span class="fc" id="L467">        int numberOfQualitiesSoFar=0;</span>
<span class="fc" id="L468">        Iterator&lt;Nucleotide&gt; iter = consensus.iterator();</span>
<span class="fc bfc" id="L469" title="All 2 branches covered.">        while(iter.hasNext()){</span>
<span class="fc" id="L470">        	Nucleotide base = iter.next();</span>
<span class="fc bfc" id="L471" title="All 2 branches covered.">            if(base.isGap()){</span>
<span class="fc" id="L472">                continue;</span>
            }
<span class="fc" id="L474">            result.append(&quot; 90&quot;);</span>
<span class="fc" id="L475">            numberOfQualitiesSoFar++;</span>
<span class="fc bfc" id="L476" title="All 2 branches covered.">            if(numberOfQualitiesSoFar%50==0){</span>
<span class="fc" id="L477">                result.append(String.format(&quot;%n&quot;));</span>
            }
<span class="fc" id="L479">        }        </span>
<span class="fc" id="L480">        writeString(String.format(&quot;BQ%n%s%n&quot;, result.toString()), out);</span>
<span class="fc" id="L481">    }</span>
   

    private static String createAssembledFromRecord(AceAssembledRead read, long fullLength){
<span class="fc" id="L485">    	IdAlignedReadInfo assembledFrom = IdAlignedReadInfo.createFrom(read, fullLength);</span>
<span class="fc" id="L486">        return String.format(&quot;AF %s %s %d%n&quot;,</span>
<span class="fc" id="L487">                assembledFrom.getId(),</span>
<span class="fc bfc" id="L488" title="All 2 branches covered.">                assembledFrom.getDirection()==Direction.FORWARD? &quot;U&quot;:&quot;C&quot;,</span>
<span class="fc" id="L489">                        assembledFrom.getStartOffset());</span>
    }
    
    
    private static String createPlacedReadRecord(AceAssembledRead read, Phd phd){
<span class="fc" id="L494">        return AceFileUtil.createAcePlacedReadRecord(</span>
<span class="fc" id="L495">                read.getId(),read,</span>
                phd, 
<span class="fc" id="L497">                read.getPhdInfo());</span>
        
    }
    private static final class IdAlignedReadInfo implements Comparable&lt;IdAlignedReadInfo&gt;{
    	private static final int TO_STRING_BUFFER_SIZE = 30;
		private final String id;
	    private final byte dir;
	    private final int startOffset;
<span class="fc" id="L505">	    private static final Direction[] DIRECTION_VALUES = Direction.values();</span>
	    public static IdAlignedReadInfo createFrom(AssembledRead read, long ungappedFullLength){
	        final Range validRange;
<span class="fc" id="L508">	        Direction dir = read.getDirection();</span>
<span class="fc" id="L509">	        Range readValidRange = read.getReadInfo().getValidRange();</span>
<span class="fc bfc" id="L510" title="All 2 branches covered.">	        if(dir==Direction.REVERSE){</span>
<span class="fc" id="L511">	            validRange = AssemblyUtil.reverseComplementValidRange(readValidRange, ungappedFullLength);</span>
	        }
	        else{
<span class="fc" id="L514">	            validRange = readValidRange;</span>
	        }
<span class="fc" id="L516">	        return new IdAlignedReadInfo(read.getId(), </span>
<span class="fc" id="L517">	                (int)(read.getGappedStartOffset()-validRange.getBegin()+1),dir);</span>
	    }
	    
	    
<span class="fc" id="L521">		private IdAlignedReadInfo(String id, int startOffset, Direction dir) {</span>
<span class="fc" id="L522">			this.id = id;</span>
<span class="fc" id="L523">			this.dir = (byte)dir.ordinal();</span>
<span class="fc" id="L524">			this.startOffset = startOffset;</span>
<span class="fc" id="L525">		}</span>


		@Override
	    public int hashCode() {
<span class="nc" id="L530">	        final int prime = 31;</span>
<span class="nc" id="L531">	        int result = 1;</span>
<span class="nc" id="L532">	        result = prime * result + id.hashCode();</span>
<span class="nc" id="L533">	        return result;</span>
	    }
	    @Override
	    public boolean equals(Object obj) {
<span class="nc bnc" id="L537" title="All 2 branches missed.">	        if (this == obj){</span>
<span class="nc" id="L538">	            return true;</span>
	        }
<span class="nc bnc" id="L540" title="All 2 branches missed.">	        if (obj == null){</span>
<span class="nc" id="L541">	            return false;</span>
	        }
<span class="nc bnc" id="L543" title="All 2 branches missed.">	        if (!(obj instanceof IdAlignedReadInfo)){</span>
<span class="nc" id="L544">	            return false;</span>
	        }
<span class="nc" id="L546">	        IdAlignedReadInfo other = (IdAlignedReadInfo) obj;</span>
<span class="nc" id="L547">	        return id.equals(other.getId());</span>
	    }
	    public String getId() {
<span class="fc" id="L550">	        return id;</span>
	    }

	    public int getStartOffset() {
<span class="fc" id="L554">	        return startOffset;</span>
	    }
	    
	    public Direction getDirection(){
<span class="fc" id="L558">	        return DIRECTION_VALUES[dir];</span>
	    }
	    @Override
	    public String toString() {
<span class="nc" id="L562">	        StringBuilder builder = new StringBuilder(TO_STRING_BUFFER_SIZE);</span>
<span class="nc" id="L563">	        builder.append(id).append(' ')</span>
<span class="nc" id="L564">	        		.append(startOffset)</span>
<span class="nc" id="L565">	        		.append(&quot;is complemented? &quot;)</span>
<span class="nc bnc" id="L566" title="All 2 branches missed.">	        		.append(getDirection() ==Direction.REVERSE);</span>
<span class="nc" id="L567">	        return builder.toString();</span>
	    }
	    /**
	    * Compares two AssembledFrom instances and compares them based on start offset
	    * then by Id.  This should match the order of AssembledFrom records 
	    * (and reads) in an .ace file.
	    */
	    @Override
	    public int compareTo(IdAlignedReadInfo o) {
<span class="fc" id="L576">	    	return JillionUtil.compare(getStartOffset(), o.getStartOffset());	    	</span>
	    }
    	
    }
    
    /**
     * Compute the consensus quality sequence as computed by the same algorithm consed uses.
     * 
     * @param contig the contig to compute the consensus qualities for; can not be null.
     * ConsensusQualities will only represent the qualities for the UNGAPPED 
     * consensus sequence.
     * @param readQualities a {@link QualitySequenceDataStore} containing
     * {@link QualitySequence}s for the underlying reads;
     * can not be null.
     * 
     * @return a {@link QualitySequence} can not be null.
     * @throws DataStoreException if there is a problem fetching quality data from 
     * the {@link QualitySequenceDataStore}.
     * @throws NullPointerException if either parameter is null.
     */
    public static QualitySequence computeConsensusQualities(Contig&lt;? extends AssembledRead&gt; contig, QualitySequenceDataStore readQualities) throws DataStoreException{
<span class="fc" id="L597">    	return ConsedConsensusQualityComputer.computeConsensusQualities(contig, readQualities);</span>
    }
    
    
    
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.5.201505241946</span></div></body></html>