<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../.resources/report.css" type="text/css"/><link rel="shortcut icon" href="../.resources/report.gif" type="image/gif"/><title>FastqUtil.java</title><link rel="stylesheet" href="../.resources/prettify.css" type="text/css"/><script type="text/javascript" src="../.resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../.sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">Jillion</a> &gt; <a href="index.source.html" class="el_package">org.jcvi.jillion.trace.fastq</a> &gt; <span class="el_source">FastqUtil.java</span></div><h1>FastqUtil.java</h1><pre class="source lang-java linenums">/*******************************************************************************
 * Jillion development code
 * 
 * This code may be freely distributed and modified under the
 * terms of the GNU Lesser General Public License.  This should
 * be distributed with the code.  If you do not have a copy,
 *  see:
 * 
 *          http://www.gnu.org/copyleft/lesser.html
 * 
 * 
 * Copyright for this code is held jointly by the individual authors.  These should be listed in the @author doc comments.
 * 
 * Information about Jillion can be found on its homepage
 * 
 *         http://jillion.sourceforge.net
 * 
 * Contributors:
 *     Danny Katzel - initial API and implementation
 ******************************************************************************/
package org.jcvi.jillion.trace.fastq;

import java.io.File;
import java.io.IOException;
import java.util.regex.Pattern;

import org.jcvi.jillion.core.io.InputStreamSupplier;
import org.jcvi.jillion.core.qual.QualitySequence;

/**
 * {@code FastqUtil} is a utility class for working with 
 * FASTQ data.
 * @author dkatzel
 *
 *
 */
public final class FastqUtil {


    /**
     * This is the {@link Pattern} to parse
     * the sequence record defline of a FASTQ record.
     * Group 1 will be the read id
     * Group 3 will be the optional comment if there is one,
     * or null if there isn't a comment.
     */
<span class="fc" id="L47">    public static final Pattern SEQ_DEFLINE_PATTERN = Pattern.compile(&quot;^@(\\S+)(\\s+)?(.+$)?&quot;);</span>
    

<span class="nc" id="L50">	private FastqUtil(){</span>
		//can not instantiate
<span class="nc" id="L52">	}</span>
    
    /**
     * Attempts to guess the {@link FastqQualityCodec} used to encode
     * the qualities in the given fastq file. 
     * This method works by analyzing the encoded quality value ranges for each
     * read getting inspected.  Since all the different quality encodings have
     * overlapping values the {@link FastqQualityCodec} used for the entire file
     * can not always be determined from just analyzing a few reads.
     * Therefore, this method might have to parse
     * ALL records in the fastq file to guarantee that the
     * correct {@link FastqQualityCodec} is returned.
     * @param fastqFile a fastq File, must exist and can not be null.
     * @return an instance of {@link FastqQualityCodec} which is the one
     * that is most likely able to decode the quality values for all the reads
     * in the file (will never be null).
     * @throws IOException if there is a problem reading the file.
     * @throws NullPointerException if fastqFile is null.
     * @throws IllegalArgumentException if the given quality string is empty
     * or if it contains any characters out of range of any known
     * quality encoding formats.
     * @throws IllegalStateException if the fastq file does not contain 
     * any records.
     */
    public static FastqQualityCodec guessQualityCodecUsed(File fastqFile) throws IOException{
<span class="fc" id="L77">    	return guessQualityCodecUsed(fastqFile, Integer.MAX_VALUE);</span>
    }
    
    /**
     * Attempts to guess the {@link FastqQualityCodec} used to encode
     * the qualities in the given fastq file. 
     * This method works by analyzing the encoded quality value ranges for each
     * read getting inspected.  Since all the different quality encodings have
     * overlapping values the {@link FastqQualityCodec} used for the entire file
     * can not always be determined from just analyzing a few reads.
     * Therefore, this method must
     * parse many (probably hundreds) of reads in the fastq file in order to get
     * a large enough sampling size to confidently return a {@link FastqQualityCodec}.
     * This method will keep looking at reads in the fastq file until
     * either the file ends, or it has looked at {@code numberOfReadsToInspect}.
     * @param fastqFile a fastq File, must exist and can not be null.
     * @param numberOfReadsToInspect the number of reads in the file to analyze.
     * 
     * @return an instance of {@link FastqQualityCodec} which is the one
     * that is most likely able to decode the quality values for all the reads
     * in the file (will never be null).
     * @throws IOException if there is a problem reading the file.
     * @throws NullPointerException if fastqFile is null.
     * @throws IllegalArgumentException if the given quality string is empty
     * or if it contains any characters out of range of any known
     * quality encoding formats.
     * @throws IllegalArgumentException if numberOfReadsToInspect is &lt; 1.
     * @throws IllegalStateException if the fastq file does not contain 
     * any records.
     */
    public static FastqQualityCodec guessQualityCodecUsed(File fastqFile, int numberOfReadsToInspect) throws IOException{
<span class="pc bpc" id="L108" title="1 of 2 branches missed.">    	if(numberOfReadsToInspect &lt;1){</span>
<span class="nc" id="L109">    		throw new IllegalArgumentException(&quot;number of reads to inspect must be &gt;=1&quot;);</span>
    	}
<span class="fc" id="L111">    	FastqQualityCodecDetectorVisitor detectorVisitor =new FastqQualityCodecDetectorVisitor(numberOfReadsToInspect);</span>
<span class="fc" id="L112">    	FastqFileParser.create(InputStreamSupplier.forFile(fastqFile), false,true, false).parse(detectorVisitor);</span>
<span class="fc" id="L113">    	return detectorVisitor.getDetectedCodec();</span>
    }
    /**
     * Attempts to guess the {@link FastqQualityCodec} used to encode
     * the qualities in the given fastq file. 
     * This method works by analyzing the encoded quality value ranges for each
     * read getting inspected.  Since all the different quality encodings have
     * overlapping values the {@link FastqQualityCodec} used for the entire file
     * can not always be determined from just analyzing a few reads.
     * Therefore, this method might have to parse
     * ALL records in the fastq file to guarantee that the
     * correct {@link FastqQualityCodec} is returned.
     * @param parser a {@link FastqParser} instance to parse; must exist and can not be null.
     * @return an instance of {@link FastqQualityCodec} which is the one
     * that is most likely able to decode the quality values for all the reads
     * in the file (will never be null).
     * @throws IOException if there is a problem reading the file.
     * @throws NullPointerException if fastqFile is null.
     * @throws IllegalArgumentException if the given quality string is empty
     * or if it contains any characters out of range of any known
     * quality encoding formats.
     * @throws IllegalStateException if the fastq file does not contain 
     * any records.
     */
    public static FastqQualityCodec guessQualityCodecUsed(FastqParser parser) throws IOException{
<span class="fc" id="L138">    	return guessQualityCodecUsed(parser, Integer.MAX_VALUE);</span>
    }
    
    /**
     * Attempts to guess the {@link FastqQualityCodec} used to encode
     * the qualities in the given fastq file. 
     * This method works by analyzing the encoded quality value ranges for each
     * read getting inspected.  Since all the different quality encodings have
     * overlapping values the {@link FastqQualityCodec} used for the entire file
     * can not always be determined from just analyzing a few reads.
     * Therefore, this method must
     * parse many (probably hundreds) of reads in the fastq file in order to get
     * a large enough sampling size to confidently return a {@link FastqQualityCodec}.
     * This method will keep looking at reads in the fastq file until
     * either the file ends, or it has looked at {@code numberOfReadsToInspect}.
     * @param parser a {@link FastqParser} instance to parse; must exist and can not be null.
     * @param numberOfReadsToInspect the number of reads in the file to analyze.
     * 
     * @return an instance of {@link FastqQualityCodec} which is the one
     * that is most likely able to decode the quality values for all the reads
     * in the file (will never be null).
     * @throws IOException if there is a problem reading the file.
     * @throws NullPointerException if fastqFile is null.
     * @throws IllegalArgumentException if the given quality string is empty
     * or if it contains any characters out of range of any known
     * quality encoding formats.
     * @throws IllegalArgumentException if numberOfReadsToInspect is &lt; 1.
     * @throws IllegalStateException if the fastq file does not contain 
     * any records.
     */
    public static FastqQualityCodec guessQualityCodecUsed(FastqParser parser, int numberOfReadsToInspect) throws IOException{
<span class="pc bpc" id="L169" title="1 of 2 branches missed.">    	if(numberOfReadsToInspect &lt;1){</span>
<span class="nc" id="L170">    		throw new IllegalArgumentException(&quot;number of reads to inspect must be &gt;=1&quot;);</span>
    	}
<span class="fc" id="L172">    	FastqQualityCodecDetectorVisitor detectorVisitor =new FastqQualityCodecDetectorVisitor(numberOfReadsToInspect);</span>
<span class="fc" id="L173">    	parser.parse(detectorVisitor);</span>
<span class="fc" id="L174">    	return detectorVisitor.getDetectedCodec();</span>
    }
    /**
     * Attempts to guess the {@link FastqQualityCodec} used to encode
     * the given qualities.
     * @param encodedQualities a String of fastq encoded qualities
     * @return an instance of {@link FastqQualityCodec} which could have been 
     * used to encode the given qualities; or {@code null} if encodedQualities
     * is empty.
     * @throws NullPointerException if the given qualities are null.
     * @throws IllegalArgumentException if the given quality string
     * contains any characters out of range of any known
     * quality encoding formats.
     */
    static FastqQualityCodec guessQualityCodecUsed(String encodedQualities){
<span class="fc bfc" id="L189" title="All 2 branches covered.">    	if(encodedQualities.isEmpty()){</span>
<span class="fc" id="L190">    		return null;</span>
    	}
    	//sanger uses 33 as an offset so any ascii values around there will 
    	//automatically be sanger
<span class="fc" id="L194">    	int sangerOffset = FastqQualityCodec.SANGER.getOffset();</span>
    	//solexa and illumina encoding have offsets of 64 so they look very similar
    	//except solexa uses a different log scale and can actually have scores as low
    	//as -5 (ascii value 59) so any values from ascii 59-63 mean solexa
<span class="fc" id="L198">    	boolean hasSolexaOnlyValues=false;</span>
<span class="fc" id="L199">    	int maxQuality=Integer.MIN_VALUE;</span>
<span class="fc" id="L200">    	int minQuality = Integer.MAX_VALUE;</span>
    	
    	//convert to char[] as optimization
    	//so we don't have to do 2x the boundary checks
    	//in String.charAt(i) 
    	//which does extra boundary checking to make
    	//sure i is in range before accessing
    	//the internal array.
<span class="fc" id="L208">    	char[] chars = encodedQualities.toCharArray();</span>
    	
<span class="fc bfc" id="L210" title="All 2 branches covered.">    	for(int i=0; i&lt;chars.length; i++){</span>
<span class="fc" id="L211">    		int asciiValue = chars[i];</span>
<span class="pc bpc" id="L212" title="1 of 2 branches missed.">    		if(asciiValue &lt;sangerOffset){</span>
<span class="nc" id="L213">    			throw new IllegalArgumentException(</span>
<span class="nc" id="L214">    					String.format(</span>
    							&quot;invalid encoded qualities has out of range ascii value %d : '%s'&quot;, 
<span class="nc" id="L216">    							asciiValue,</span>
    							encodedQualities));
    		}
<span class="fc bfc" id="L219" title="All 2 branches covered.">    		if(asciiValue &lt;59){</span>
<span class="fc" id="L220">    			return FastqQualityCodec.SANGER;</span>
    		}
<span class="fc bfc" id="L222" title="All 2 branches covered.">    		if(asciiValue &lt; 64){</span>
<span class="fc" id="L223">    			hasSolexaOnlyValues=true;</span>
    		}
<span class="fc bfc" id="L225" title="All 2 branches covered.">    		if(asciiValue &gt; maxQuality){</span>
<span class="fc" id="L226">    			maxQuality = asciiValue;</span>
    		}
<span class="fc bfc" id="L228" title="All 2 branches covered.">    		if(asciiValue &lt; minQuality){</span>
<span class="fc" id="L229">    			minQuality = asciiValue;</span>
    		}
    	}
    	//if we get this far, we don't have 
    	//any QVs ascii values &lt; 59
    	//we could still be sanger
    	//and only have high quality qvs
    	//(possible but unlikely outside of testing)
    	//or only low super low illumina/solexa qvs
    	//most qvs are between 1-40 so check to see if we have
    	//any values beyond that.
    	
<span class="fc bfc" id="L241" title="All 2 branches covered.">		if(maxQuality &lt;  sangerOffset + 41){</span>
			//we have QVs up to sanger 40
			//or illumina qv 9
			//assume sanger
<span class="fc" id="L245">			return FastqQualityCodec.SANGER;</span>
		}
    	
<span class="fc bfc" id="L248" title="All 2 branches covered.">    	if(hasSolexaOnlyValues){</span>
<span class="fc" id="L249">    		return FastqQualityCodec.SOLEXA;</span>
    	}
    	//if we get here then we only saw encoded values from ascii 64 +
    	//assume illumina, solexa scaling is so close at good quality anyway
    	//that I don't think it matters.
    	
<span class="fc" id="L255">    	return FastqQualityCodec.ILLUMINA;</span>
    }
    /**
     * Inner class that visits the first X reads of a fastq file
     * and notes the {@link FastqQualityCodec} that can be used.
     * After a sample size of X reads is analyzed, the most frequent
     * codec is returned.
     * @author dkatzel
     *
     */
    private static final class FastqQualityCodecDetectorVisitor implements FastqVisitor{
<span class="fc" id="L266">    	private int numberOfRecordsVisited=0;</span>
    	private final int maxNumberOfRecordsToVisit;
		
<span class="fc" id="L269">    	FastqQualityCodecDetectorRecordVisitor recordVisitor = new FastqQualityCodecDetectorRecordVisitor();</span>
    	
<span class="fc" id="L271">		public FastqQualityCodecDetectorVisitor(int maxNumberOfRecordsToVisit) {</span>
<span class="fc" id="L272">			this.maxNumberOfRecordsToVisit = maxNumberOfRecordsToVisit;</span>
<span class="fc" id="L273">		}</span>
		public FastqQualityCodec getDetectedCodec() throws MixedFastqEncodings {
<span class="fc" id="L275">			return recordVisitor.getMostFrequentCodec();	</span>
		}
		@Override
		public FastqRecordVisitor visitDefline(FastqVisitorCallback callback,
				String id, String optionalComment) {
<span class="pc bpc" id="L280" title="1 of 2 branches missed.">			if(numberOfRecordsVisited&lt;maxNumberOfRecordsToVisit){</span>
<span class="fc" id="L281">				numberOfRecordsVisited++;</span>
<span class="fc" id="L282">				return recordVisitor;</span>
			}
<span class="nc" id="L284">			callback.haltParsing();</span>
<span class="nc" id="L285">			return null;</span>
		}
		/**
		 * This method will be the most frequently
		 * one called since most fastq files
		 * are huge so {@link org.jcvi.jillion.trace.fastq.FastqVisitor.FastqVisitorCallback#haltParsing()} will get called
		 * 
		 */
		@Override
		public void halted() {
			//no-op
<span class="nc" id="L296">		}</span>
		/**
		 * It is possible that the fastq
		 * file is so small that 
		 * {@link org.jcvi.jillion.trace.fastq.FastqVisitor.FastqVisitorCallback#haltParsing()}
		 * will not get called.
		 * 
		 */
		@Override
		public void visitEnd() {			
			//no-op
<span class="fc" id="L307">		}</span>
    	
    }
    
<span class="fc" id="L311">    private static final class FastqQualityCodecDetectorRecordVisitor implements FastqRecordVisitor{</span>
<span class="fc" id="L312">    	private long numSanger=0;</span>
<span class="fc" id="L313">    	private long numIllumina=0;</span>
<span class="fc" id="L314">    	private long numPossiblySolexa=0;</span>
<span class="fc" id="L315">    	private long numEither=0;</span>
    	
<span class="fc" id="L317">    	private int overAllMin= Integer.MAX_VALUE;</span>
<span class="fc" id="L318">    	private int overallMax = Integer.MIN_VALUE;</span>
		@Override
		public void visitNucleotides(String nucleotides) {
			//no-op			
<span class="fc" id="L322">		}</span>

		@Override
		public void visitEncodedQualities(String encodedQualities) {
		
			
<span class="fc" id="L328">	    	boolean hasSolexaOnlyValues=false;</span>
<span class="fc" id="L329">	    	int maxQuality=Integer.MIN_VALUE;</span>

	    	//convert to char[] as optimization
	    	//so we don't have to do 2x the boundary checks
	    	//in String.charAt(i) 
	    	//which does extra boundary checking to make
	    	//sure i is in range before accessing
	    	//the internal array.
<span class="fc" id="L337">	    	char[] chars = encodedQualities.toCharArray();</span>
<span class="fc bfc" id="L338" title="All 2 branches covered.">	    	for(int i=0; i&lt;chars.length; i++){</span>
<span class="fc" id="L339">	    		int asciiValue = chars[i];</span>
<span class="pc bpc" id="L340" title="1 of 2 branches missed.">	    		if(asciiValue &lt;33){</span>
<span class="nc" id="L341">	    			throw new IllegalArgumentException(</span>
<span class="nc" id="L342">	    					String.format(</span>
	    							&quot;invalid encoded qualities has out of range ascii value %d : '%s'&quot;, 
<span class="nc" id="L344">	    							asciiValue,</span>
	    							encodedQualities));
	    		}
<span class="fc bfc" id="L347" title="All 2 branches covered.">	    		if(asciiValue &gt; maxQuality){</span>
<span class="fc" id="L348">	    		    maxQuality = asciiValue;</span>
	    		}
<span class="fc bfc" id="L350" title="All 2 branches covered.">    	    		if(asciiValue &gt; overallMax){</span>
<span class="fc" id="L351">    	    		    overallMax = asciiValue;</span>
                        }
<span class="fc bfc" id="L353" title="All 2 branches covered.">                        if(asciiValue &lt; overAllMin){</span>
<span class="fc" id="L354">                            overAllMin = asciiValue;</span>
                        }
	    		//sanger uses 33 as an offset so any ascii values around there will 
		    	//automatically be sanger
<span class="fc bfc" id="L358" title="All 2 branches covered.">	    		if(asciiValue &lt;59){</span>
<span class="fc" id="L359">	    			numSanger++;</span>
<span class="fc" id="L360">	    			return;</span>
	    		}
	    		//solexa and illumina encoding have offsets of 64 so they look very similar
		    	//except solexa uses a different log scale and can actually have scores as low
		    	//as -5 (ascii value 59) so any values from ascii 59-63 mean solexa
<span class="fc bfc" id="L365" title="All 2 branches covered.">	    		if(asciiValue &lt; 64){</span>
<span class="fc" id="L366">	    			hasSolexaOnlyValues=true;</span>
	    		}
	    		
	    	}
	    	//a quality more than this is highly improbable
	    	//for sanger encoding so assume illumina or solexa
<span class="fc bfc" id="L372" title="All 2 branches covered.">	    	if(maxQuality &gt; 80){</span>
<span class="pc bpc" id="L373" title="1 of 2 branches missed.">	    		if(hasSolexaOnlyValues){</span>
<span class="nc" id="L374">	    			numPossiblySolexa++;</span>
	    		}else{
<span class="fc" id="L376">	    			numIllumina ++;</span>
	    		}
	    	}else{
	    		//could not tell which encoding
	    		//this could be a high quality sanger 
	    		//or low quality illumina read
<span class="fc" id="L382">	    		numEither++;</span>
	    	}
	    	
<span class="fc" id="L385">		}</span>

		
		
		@Override
		public void visitEnd() {
			//no-op			
<span class="fc" id="L392">		}</span>
    	
		@Override
		public void halted() {
			//no-op			
<span class="nc" id="L397">		}</span>
		
		public FastqQualityCodec getMostFrequentCodec() throws MixedFastqEncodings{
<span class="pc bpc" id="L400" title="2 of 8 branches missed.">			if(numSanger==0 &amp;&amp; numPossiblySolexa==0 &amp;&amp; numIllumina==0 &amp;&amp; numEither ==0){</span>
<span class="nc" id="L401">				throw new IllegalStateException(&quot;fastq file must not be empty&quot;);</span>
			}
<span class="pc bpc" id="L403" title="1 of 4 branches missed.">			if(numSanger&gt;0 &amp;&amp; numIllumina&gt;0){</span>
				//mix could be invalid upstream process
				//that mixed the encodings together.
				//This has happened where many 
				//fastq files are processed differently 
				//but then combined into one large fastq file
<span class="nc" id="L409">				throw new MixedFastqEncodings(numSanger, numPossiblySolexa, numIllumina, numEither);</span>
			}		
			//if our file has anything that
			//is definitely sanger encoded then we 
			//must be sanger encoded.
<span class="fc bfc" id="L414" title="All 2 branches covered.">			if(numSanger &gt;0){</span>
<span class="fc" id="L415">				return FastqQualityCodec.SANGER;</span>
			}
			//we could be either high quality sanger
			//or low quality illumina.
			//we have 0 def sanger and 0 def illumina
			//AND
			//we have QVs up to sanger 40
			//or illumina qv 9:
			//assume sanger
<span class="pc bpc" id="L424" title="1 of 4 branches missed.">			if(numIllumina==0 &amp;&amp; overallMax &lt;  FastqQualityCodec.SANGER.getOffset() + 41){</span>
<span class="fc" id="L425">				return FastqQualityCodec.SANGER;				</span>
			}
			
			//if we get this far than we use a 64 offset
			//but don't yet know if we are solexa or illumina
<span class="pc bpc" id="L430" title="1 of 2 branches missed.">			if(numPossiblySolexa&gt;0){</span>
<span class="nc" id="L431">				return FastqQualityCodec.SOLEXA;</span>
			}
<span class="fc" id="L433">			return FastqQualityCodec.ILLUMINA;</span>
		}

		@Override
		public void visitQualities(QualitySequence qualities) {
			//no-op since we there is no encoding.
			
<span class="nc" id="L440">		}</span>
    }
    
    private static class MixedFastqEncodings extends IOException{
		private static final long serialVersionUID = 1L;

		MixedFastqEncodings(long numSanger, long numSolexa, long numIllumina, long numEither){
<span class="nc" id="L447">    		super(String.format(&quot;many reads are encoded differently #sanger =%d #illuminia = %d #either = %d&quot; , numSanger,numIllumina, numSolexa+numEither));</span>
<span class="nc" id="L448">    	}</span>
    }
   
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.5.201505241946</span></div></body></html>