<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../.resources/report.css" type="text/css"/><link rel="shortcut icon" href="../.resources/report.gif" type="image/gif"/><title>DefaultContig.java</title><link rel="stylesheet" href="../.resources/prettify.css" type="text/css"/><script type="text/javascript" src="../.resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../.sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">Jillion</a> &gt; <a href="index.source.html" class="el_package">org.jcvi.jillion.internal.assembly</a> &gt; <span class="el_source">DefaultContig.java</span></div><h1>DefaultContig.java</h1><pre class="source lang-java linenums">/*******************************************************************************
 * Jillion development code
 * 
 * This code may be freely distributed and modified under the
 * terms of the GNU Lesser General Public License.  This should
 * be distributed with the code.  If you do not have a copy,
 *  see:
 * 
 *          http://www.gnu.org/copyleft/lesser.html
 * 
 * 
 * Copyright for this code is held jointly by the individual authors.  These should be listed in the @author doc comments.
 * 
 * Information about Jillion can be found on its homepage
 * 
 *         http://jillion.sourceforge.net
 * 
 * Contributors:
 *     Danny Katzel - initial API and implementation
 ******************************************************************************/
/*
 * Created on Jan 15, 2009
 *
 * @author dkatzel
 */
package org.jcvi.jillion.internal.assembly;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;

import org.jcvi.jillion.assembly.AbstractContigBuilder;
import org.jcvi.jillion.assembly.AssembledRead;
import org.jcvi.jillion.assembly.AssembledReadBuilder;
import org.jcvi.jillion.assembly.Contig;
import org.jcvi.jillion.core.Direction;
import org.jcvi.jillion.core.Range;
import org.jcvi.jillion.core.io.IOUtil;
import org.jcvi.jillion.core.residue.nt.NucleotideSequence;
import org.jcvi.jillion.core.residue.nt.NucleotideSequenceBuilder;
import org.jcvi.jillion.core.util.MapUtil;
import org.jcvi.jillion.core.util.ThrowingStream;
import org.jcvi.jillion.core.util.iter.IteratorUtil;
import org.jcvi.jillion.core.util.iter.StreamingIterator;

public final class DefaultContig&lt;T extends AssembledRead&gt; implements Contig&lt;T&gt;{

	private final NucleotideSequence consensus;
    private final String id;
    private final Map&lt;String, T&gt; mapById;
    
<span class="fc" id="L53">    public DefaultContig(String id, NucleotideSequence consensus, Set&lt;T&gt; assembledReads){</span>
<span class="pc bpc" id="L54" title="1 of 2 branches missed.">    	if(id==null){</span>
<span class="nc" id="L55">    		throw new NullPointerException(&quot;id can not be null&quot;);</span>
    	}
<span class="pc bpc" id="L57" title="1 of 2 branches missed.">    	if(consensus==null){</span>
<span class="nc" id="L58">    		throw new NullPointerException(&quot;consensus can not be null&quot;);</span>
    	}
<span class="fc" id="L60">        this.id = id;</span>
<span class="fc" id="L61">        this.consensus = consensus;</span>
<span class="fc" id="L62">        int capacity = MapUtil.computeMinHashMapSizeWithoutRehashing(assembledReads.size());</span>
<span class="fc" id="L63">        mapById = new LinkedHashMap&lt;String, T&gt;(capacity);</span>
<span class="fc bfc" id="L64" title="All 2 branches covered.">        for(T r : assembledReads){</span>
<span class="fc" id="L65">            mapById.put(r.getId(), r);</span>
<span class="fc" id="L66">        }</span>
       
<span class="fc" id="L68">    }</span>
<span class="fc" id="L69">    public DefaultContig(String id, NucleotideSequence consensus, Map&lt;String, T&gt; assembledReads){</span>
<span class="pc bpc" id="L70" title="1 of 2 branches missed.">    	if(id==null){</span>
<span class="nc" id="L71">    		throw new NullPointerException(&quot;id can not be null&quot;);</span>
    	}
<span class="pc bpc" id="L73" title="1 of 2 branches missed.">    	if(consensus==null){</span>
<span class="nc" id="L74">    		throw new NullPointerException(&quot;consensus can not be null&quot;);</span>
    	}
<span class="pc bpc" id="L76" title="1 of 2 branches missed.">    	if(assembledReads ==null){</span>
<span class="nc" id="L77">    		throw new NullPointerException(&quot;assembled read map can not be null&quot;);</span>
    	}
<span class="fc" id="L79">        this.id = id;</span>
<span class="fc" id="L80">        this.consensus = consensus;</span>
<span class="fc" id="L81">        mapById = assembledReads;</span>
       
<span class="fc" id="L83">    }</span>

    @Override
    public NucleotideSequence getConsensusSequence() {
<span class="fc" id="L87">        return consensus;</span>
    }

    @Override
    public String getId() {
<span class="fc" id="L92">        return id;</span>
    }

    @Override
    public long getNumberOfReads() {
<span class="fc" id="L97">        return mapById.size();</span>
    }
    @Override
    public T getRead(String id) {
<span class="fc" id="L101">        return mapById.get(id);</span>
    }
    @Override
    public StreamingIterator&lt;T&gt; getReadIterator() {       
<span class="fc" id="L105">        return IteratorUtil.createStreamingIterator(mapById.values().iterator());</span>
    }

    


    @Override
   	public ThrowingStream&lt;T&gt; reads() {
<span class="fc" id="L113">   		return ThrowingStream.asThrowingStream(mapById.values().stream());</span>
   	}
	@Override
    public boolean containsRead(String placedReadId) {
<span class="fc" id="L117">        return mapById.containsKey(placedReadId);</span>
    }

	@Override
	public int hashCode() {
<span class="nc" id="L122">		final int prime = 31;</span>
<span class="nc" id="L123">		int result = 1;</span>
<span class="nc" id="L124">		result = prime * result</span>
<span class="nc" id="L125">				+ consensus.hashCode();</span>
<span class="nc" id="L126">		result = prime * result +  id.hashCode();</span>
<span class="nc" id="L127">		result = prime * result +  mapById.hashCode();</span>
<span class="nc" id="L128">		return result;</span>
	}

	@Override
	public boolean equals(Object obj) {
<span class="nc bnc" id="L133" title="All 2 branches missed.">		if (this == obj) {</span>
<span class="nc" id="L134">			return true;</span>
		}
<span class="nc bnc" id="L136" title="All 2 branches missed.">		if (obj == null) {</span>
<span class="nc" id="L137">			return false;</span>
		}
<span class="nc bnc" id="L139" title="All 2 branches missed.">		if (!(obj instanceof Contig)) {</span>
<span class="nc" id="L140">			return false;</span>
		}
<span class="nc" id="L142">		Contig&lt;?&gt; other = (Contig&lt;?&gt;) obj;</span>
		
<span class="nc bnc" id="L144" title="All 2 branches missed.">		if (!id.equals(other.getId())) {</span>
<span class="nc" id="L145">			return false;</span>
		}
<span class="nc bnc" id="L147" title="All 2 branches missed.">		if (!consensus.equals(other.getConsensusSequence())) {</span>
<span class="nc" id="L148">			return false;</span>
		}
<span class="nc bnc" id="L150" title="All 2 branches missed.">		if (getNumberOfReads()!=other.getNumberOfReads()) {</span>
<span class="nc" id="L151">			return false;</span>
		}
<span class="nc bnc" id="L153" title="All 2 branches missed.">		for(Entry&lt;String, T&gt; entry : mapById.entrySet()){</span>
<span class="nc" id="L154">			String readId = entry.getKey();</span>
<span class="nc bnc" id="L155" title="All 2 branches missed.">			if(!other.containsRead(readId)){</span>
<span class="nc" id="L156">				return false;</span>
			}
<span class="nc bnc" id="L158" title="All 2 branches missed.">			if(!entry.getValue().equals(other.getRead(readId))){</span>
<span class="nc" id="L159">				return false;</span>
			}
<span class="nc" id="L161">		}</span>
		
<span class="nc" id="L163">		return true;</span>
	}

	public static final class Builder extends AbstractContigBuilder&lt;AssembledRead, Contig&lt;AssembledRead&gt;&gt;{
        public Builder(String id, String consensus){
<span class="fc" id="L168">           this(id, new NucleotideSequenceBuilder(consensus)</span>
<span class="fc" id="L169">           					.build());</span>
<span class="fc" id="L170">        }</span>
        
        public &lt;R extends AssembledRead, C extends Contig&lt;R&gt;&gt; Builder(C copy){
<span class="fc" id="L173">            this(copy.getId(), copy.getConsensusSequence());</span>
<span class="fc" id="L174">            StreamingIterator&lt;R&gt; iter =null;</span>
            try{
<span class="fc" id="L176">            	 iter = copy.getReadIterator();</span>
<span class="fc bfc" id="L177" title="All 2 branches covered.">            	 while(iter.hasNext()){</span>
<span class="fc" id="L178">            		 R read = iter.next();</span>
<span class="fc" id="L179">            		 addRead(read);</span>
<span class="fc" id="L180">            	 }</span>
            }finally{
<span class="pc" id="L182">            	IOUtil.closeAndIgnoreErrors(iter);</span>
<span class="fc" id="L183">            }</span>
<span class="fc" id="L184">         }</span>
        public Builder(String id, NucleotideSequence consensus){
<span class="fc" id="L186">            super(id,consensus);</span>
<span class="fc" id="L187">        }</span>
        public Builder addRead(String id, int offset,String basecalls){
<span class="fc" id="L189">            return addRead(id, offset, basecalls, Direction.FORWARD);</span>
        }
        public Builder addRead(String id, int offset,String basecalls, Direction dir){
<span class="fc" id="L192">            int numberOfGaps = computeNumberOfGapsIn(basecalls);</span>
<span class="fc" id="L193">            int ungappedLength = basecalls.length()-numberOfGaps;</span>
<span class="fc" id="L194">            return addRead(id, offset, </span>
<span class="fc" id="L195">            		Range.ofLength(ungappedLength),basecalls, </span>
            		dir,ungappedLength);
        }
        /**
         * @param basecalls
         * @return
         */
        private int computeNumberOfGapsIn(String basecalls) {
<span class="fc" id="L203">            int count=0;</span>
<span class="fc bfc" id="L204" title="All 2 branches covered.">            for(int i=0; i&lt;basecalls.length(); i++){</span>
<span class="fc bfc" id="L205" title="All 2 branches covered.">                if(basecalls.charAt(i) == '-'){</span>
<span class="fc" id="L206">                    count++;</span>
                }
            }
<span class="fc" id="L209">            return count;</span>
        }
        @Override
        public Builder addRead(String id, int offset,Range validRange, String basecalls, Direction dir, int fullUngappedLength){            
<span class="pc bpc" id="L213" title="1 of 2 branches missed.">            if(offset &lt;0){</span>
<span class="nc" id="L214">                throw new IllegalArgumentException(&quot;circular reads not supported&quot;);</span>
                
              }
<span class="fc" id="L217">            super.addRead(id, offset, validRange, basecalls, dir,fullUngappedLength);</span>
<span class="fc" id="L218">            return this;            </span>
        }
        
       
        public DefaultContig&lt;AssembledRead&gt; build(){
<span class="pc bpc" id="L223" title="1 of 2 branches missed.">        	 if(consensusCaller !=null){</span>
<span class="nc" id="L224">      			recallConsensusNow();</span>
              }
<span class="fc" id="L226">        	 int capacity = MapUtil.computeMinHashMapSizeWithoutRehashing(numberOfReads());</span>
<span class="fc" id="L227">            Map&lt;String,AssembledRead&gt; reads = new LinkedHashMap&lt;String, AssembledRead&gt;(capacity);</span>
<span class="fc" id="L228">            NucleotideSequence consensus = getConsensusBuilder().build();</span>
<span class="fc bfc" id="L229" title="All 2 branches covered.">            for(AssembledReadBuilder&lt;AssembledRead&gt; builder : getAllAssembledReadBuilders()){</span>
<span class="fc" id="L230">                reads.put(builder.getId(), builder.build(consensus));</span>
<span class="fc" id="L231">            }</span>
<span class="fc" id="L232">            return new DefaultContig&lt;AssembledRead&gt;(getContigId(), consensus, reads);</span>
        }
        /**
        * {@inheritDoc}
        */
        @Override
        protected AssembledReadBuilder&lt;AssembledRead&gt; createPlacedReadBuilder(
                AssembledRead read) {
<span class="fc" id="L240">            return DefaultAssembledRead.createBuilder(</span>
<span class="fc" id="L241">                    read.getId(), </span>
<span class="fc" id="L242">                    read.getNucleotideSequence().toString(), </span>
<span class="fc" id="L243">                    (int)read.getGappedStartOffset(), </span>
<span class="fc" id="L244">                    read.getDirection(), </span>
<span class="fc" id="L245">                    read.getReadInfo().getValidRange(), </span>
<span class="fc" id="L246">                    read.getReadInfo().getUngappedFullLength());</span>
        }
        /**
        * {@inheritDoc}
        */
        @Override
        protected AssembledReadBuilder&lt;AssembledRead&gt; createPlacedReadBuilder(
                String id, int offset, Range validRange, String basecalls,
                Direction dir, int fullUngappedLength) {
<span class="fc" id="L255">            return DefaultAssembledRead.createBuilder(</span>
                    id, 
                    basecalls, 
                    offset, 
                    dir, 
                    validRange, 
                    fullUngappedLength);
        }
    }

}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.5.201505241946</span></div></body></html>