<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml" lang="en"><head><meta http-equiv="Content-Type" content="text/html;charset=UTF-8"/><link rel="stylesheet" href="../.resources/report.css" type="text/css"/><link rel="shortcut icon" href="../.resources/report.gif" type="image/gif"/><title>Cas2Consed.java</title><link rel="stylesheet" href="../.resources/prettify.css" type="text/css"/><script type="text/javascript" src="../.resources/prettify.js"></script></head><body onload="window['PR_TAB_WIDTH']=4;prettyPrint()"><div class="breadcrumb" id="breadcrumb"><span class="info"><a href="../.sessions.html" class="el_session">Sessions</a></span><a href="../index.html" class="el_report">Jillion</a> &gt; <a href="index.source.html" class="el_package">org.jcvi.jillion.assembly.clc.cas.consed</a> &gt; <span class="el_source">Cas2Consed.java</span></div><h1>Cas2Consed.java</h1><pre class="source lang-java linenums">/*******************************************************************************
 * Jillion development code
 * 
 * This code may be freely distributed and modified under the
 * terms of the GNU Lesser General Public License.  This should
 * be distributed with the code.  If you do not have a copy,
 *  see:
 * 
 *          http://www.gnu.org/copyleft/lesser.html
 * 
 * 
 * Copyright for this code is held jointly by the individual authors.  These should be listed in the @author doc comments.
 * 
 * Information about Jillion can be found on its homepage
 * 
 *         http://jillion.sourceforge.net
 * 
 * Contributors:
 *     Danny Katzel - initial API and implementation
 ******************************************************************************/
package org.jcvi.jillion.assembly.clc.cas.consed;

import java.io.File;
import java.io.IOException;
import java.util.Date;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Map.Entry;

import org.jcvi.jillion.assembly.clc.cas.AbstractAlignedReadCasVisitor;
import org.jcvi.jillion.assembly.clc.cas.CasGappedReferenceDataStore;
import org.jcvi.jillion.assembly.clc.cas.read.CasPlacedRead;
import org.jcvi.jillion.assembly.consed.ConsedUtil;
import org.jcvi.jillion.assembly.consed.ace.AceContig;
import org.jcvi.jillion.assembly.consed.ace.AceContigBuilder;
import org.jcvi.jillion.assembly.consed.ace.AceFileWriter;
import org.jcvi.jillion.assembly.consed.ace.AceFileWriterBuilder;
import org.jcvi.jillion.assembly.consed.ace.PhdInfo;
import org.jcvi.jillion.assembly.consed.ace.WholeAssemblyAceTag;
import org.jcvi.jillion.assembly.consed.phd.Phd;
import org.jcvi.jillion.assembly.consed.phd.PhdBallWriter;
import org.jcvi.jillion.assembly.consed.phd.PhdDataStore;
import org.jcvi.jillion.assembly.consed.phd.PhdFileDataStoreBuilder;
import org.jcvi.jillion.assembly.consed.phd.PhdWriter;
import org.jcvi.jillion.assembly.util.GapQualityValueStrategy;
import org.jcvi.jillion.assembly.util.consensus.NextGenReferenceConsensusRecaller;
import org.jcvi.jillion.core.Range;
import org.jcvi.jillion.core.datastore.DataStoreException;
import org.jcvi.jillion.core.datastore.DataStoreProviderHint;
import org.jcvi.jillion.core.io.IOUtil;
import org.jcvi.jillion.core.qual.PhredQuality;
import org.jcvi.jillion.core.qual.QualitySequenceDataStore;
import org.jcvi.jillion.core.residue.nt.NucleotideSequence;
import org.jcvi.jillion.core.util.DateUtil;
import org.jcvi.jillion.core.util.iter.StreamingIterator;
import org.jcvi.jillion.fasta.nt.NucleotideFastaDataStore;
import org.jcvi.jillion.fasta.nt.NucleotideFastaFileDataStoreBuilder;
import org.jcvi.jillion.trace.Trace;
import org.jcvi.jillion.trace.fastq.FastqFileDataStore;
import org.jcvi.jillion.trace.sff.SffFileDataStoreBuilder;

public class Cas2Consed extends  AbstractAlignedReadCasVisitor{

	private final Map&lt;String, AceContigBuilder&gt; contigBuilders;
	private final File consedOutputDir;
<span class="fc" id="L67">	private final Date phdDate = new Date();</span>
	
	private final PhdWriter phdOut;
	
<span class="fc" id="L71">	private File chromatDir = null;</span>
	private final File phdFile;
	
	private final String prefix;
	
	public Cas2Consed(File casFile,	CasGappedReferenceDataStore gappedReferenceDataStore, File consedOutputDir,
			String prefix) throws DataStoreException, IOException {
<span class="fc" id="L78">		super(casFile.getParentFile(), gappedReferenceDataStore);</span>
		
<span class="pc bpc" id="L80" title="1 of 2 branches missed.">		if(consedOutputDir ==null){</span>
<span class="nc" id="L81">			throw new NullPointerException(&quot;output dir can not be null&quot;);</span>
		}
<span class="pc bpc" id="L83" title="1 of 2 branches missed.">		if(prefix == null){</span>
<span class="nc" id="L84">			throw new NullPointerException(&quot;prefix can not be null&quot;);</span>
		}
<span class="fc" id="L86">		this.prefix = prefix.trim();</span>
<span class="pc bpc" id="L87" title="1 of 2 branches missed.">		if(this.prefix.isEmpty()){</span>
<span class="nc" id="L88">			throw new IllegalArgumentException(&quot;prefix must contain non-whitespace&quot;);</span>
		}
<span class="fc" id="L90">		contigBuilders = new LinkedHashMap&lt;String, AceContigBuilder&gt;();</span>
<span class="fc" id="L91">		this.consedOutputDir = consedOutputDir;</span>
		
<span class="fc" id="L93">		StreamingIterator&lt;String&gt; referenceIdIterator = null;</span>
<span class="fc" id="L94">		StreamingIterator&lt;NucleotideSequence&gt; referenceSequenceIterator = null;</span>
		
<span class="fc" id="L96">		referenceIdIterator = gappedReferenceDataStore.idIterator();</span>
<span class="fc" id="L97">		referenceSequenceIterator = gappedReferenceDataStore.iterator();</span>
		
		try{
<span class="fc bfc" id="L100" title="All 2 branches covered.">			while(referenceIdIterator.hasNext()){</span>
<span class="fc" id="L101">				String id = referenceIdIterator.next();</span>
<span class="fc" id="L102">				NucleotideSequence seq = referenceSequenceIterator.next();</span>
<span class="fc" id="L103">				contigBuilders.put(id, new AceContigBuilder(id, seq));</span>
<span class="fc" id="L104">			}</span>
		}finally{
<span class="pc" id="L106">			IOUtil.closeAndIgnoreErrors(referenceIdIterator, referenceSequenceIterator);</span>
<span class="fc" id="L107">		}</span>
<span class="fc" id="L108">		File phdDir = new File(consedOutputDir, &quot;phdball_dir&quot;);</span>
		
<span class="fc" id="L110">		phdFile = new File(phdDir, &quot;phd.ball.1&quot;);</span>
<span class="fc" id="L111">		IOUtil.mkdirs(phdDir);</span>
<span class="fc" id="L112">		phdOut = new PhdBallWriter(phdFile);</span>
<span class="fc" id="L113">	}</span>

	public final File getPhdBallFile() {
<span class="nc" id="L116">		return phdFile;</span>
	}

	public final void setChromatDir(File chromatDir) {
<span class="nc" id="L120">		this.chromatDir = chromatDir;</span>
<span class="nc" id="L121">	}</span>

	@Override
	protected void notAligned(Trace currentTrace) {
		//no-op
		//we don't care about reads that don't align
<span class="fc" id="L127">	}</span>

	@Override
	protected final void aligned(Trace traceOfRead, String referenceId,
			CasPlacedRead read) {
<span class="fc" id="L132">		AceContigBuilder builder = contigBuilders.get(referenceId);</span>
<span class="pc bpc" id="L133" title="1 of 2 branches missed.">		if(!(traceOfRead instanceof PhdReadRecord)){</span>
<span class="nc" id="L134">			throw new IllegalStateException(&quot;not a valid phd record &quot; + traceOfRead);</span>
		}
<span class="fc" id="L136">		PhdReadRecord phdReadRecord = (PhdReadRecord) traceOfRead;</span>
<span class="fc" id="L137">		PhdInfo phdInfo = phdReadRecord.getPhdInfo();</span>
		
		//add read to the correct builder
<span class="fc" id="L140">		builder.addRead(read.getId(), read.getNucleotideSequence(), (int) read.getGappedStartOffset(),</span>
<span class="fc" id="L141">				read.getDirection(), read.getReadInfo().getValidRange(), phdInfo, read.getReadInfo().getUngappedFullLength());</span>
		
		//write out phd record to phdball
		try {
<span class="fc" id="L145">			phdOut.write(phdReadRecord.getPhd());</span>
<span class="nc" id="L146">		} catch (IOException e) {</span>
<span class="nc" id="L147">			throw new IllegalStateException(&quot;error writing out phd record for  &quot; + traceOfRead, e);</span>
<span class="fc" id="L148">		}</span>
<span class="fc" id="L149">		afterVisitMatch(referenceId, read,	phdReadRecord.getPhd(), phdInfo);</span>
<span class="fc" id="L150">	}</span>

	protected void afterVisitMatch(String referenceId, CasPlacedRead read,	Phd phd, PhdInfo phdInfo){
		//no-op by default
<span class="fc" id="L154">	}</span>
	
	protected void postProcess(AceContigBuilder builder){
		//no-op by default
<span class="fc" id="L158">	}</span>
	
	protected void visitAce(Range scaffoldRange, AceContig contig){
		//no-op
<span class="fc" id="L162">	}</span>
	
	@Override
	public void visitEnd() {
		//we have visited the entire cas file
		//time to write out the data
		try {
<span class="fc" id="L169">			phdOut.close();</span>
<span class="fc" id="L170">			PhdDataStore phdDataStore = new PhdFileDataStoreBuilder(phdFile)</span>
										//.hint(DataStoreProviderHint.RANDOM_ACCESS_OPTIMIZE_MEMORY)
<span class="fc" id="L172">										.build();</span>
<span class="fc" id="L173">			File editDir = new File(consedOutputDir, &quot;edit_dir&quot;);</span>
<span class="fc" id="L174">			File aceFile = new File(editDir, prefix + &quot;.ace.1&quot;);</span>
			//Explicitly make a tmp dir under the editDir
			//so we don't have to worry about space issues on grid nodes.
			//first need to create edit_dir 
			//to be able to make the temp area.
<span class="fc" id="L179">			IOUtil.mkdirs(editDir);</span>
<span class="fc" id="L180">			File tmpDir = IOUtil.createTempDir(&quot;cas2consed&quot;, null, editDir);</span>
			try{
<span class="fc" id="L182">				AceFileWriter aceWriter = new AceFileWriterBuilder(aceFile, phdDataStore)</span>
											//set tmp dir Explicitly
<span class="fc" id="L184">											.tmpDir(tmpDir)</span>
<span class="fc" id="L185">											.build();</span>
<span class="fc" id="L186">				Iterator&lt;Entry&lt;String, AceContigBuilder&gt;&gt; referenceEntryIter = contigBuilders.entrySet().iterator();</span>
<span class="fc" id="L187">				QualitySequenceDataStore qualityDataStore =  phdDataStore.asQualityDataStore();</span>
<span class="fc bfc" id="L188" title="All 2 branches covered.">				while(referenceEntryIter.hasNext()){</span>
<span class="fc" id="L189">					Entry&lt;String, AceContigBuilder&gt; refEntry = referenceEntryIter.next();</span>
<span class="fc" id="L190">					AceContigBuilder contigBuilder = refEntry.getValue();</span>
					
<span class="fc" id="L192">					contigBuilder.recallConsensus(new NextGenReferenceConsensusRecaller(), qualityDataStore, GapQualityValueStrategy.LOWEST_FLANKING);</span>
<span class="fc" id="L193">					postProcess(contigBuilder);</span>
<span class="fc" id="L194">					visitBeginReference(refEntry.getKey());</span>
<span class="fc bfc" id="L195" title="All 2 branches covered.">					for(Entry&lt;Range,AceContig&gt; entry : ConsedUtil.split0xContig(contigBuilder,true).entrySet()){</span>
<span class="fc" id="L196">	                    AceContig splitContig = entry.getValue();</span>
	                    
<span class="fc" id="L198">	                    visitAce(entry.getKey(), splitContig);</span>
	                    
<span class="fc" id="L200">						aceWriter.write(splitContig);</span>
<span class="fc" id="L201">	                }</span>
<span class="fc" id="L202">					visitEndReference();</span>
					//allows seen builders to be garbage collected if needed
<span class="fc" id="L204">					referenceEntryIter.remove();</span>
<span class="fc" id="L205">				}</span>
<span class="fc" id="L206">				aceWriter.write(new WholeAssemblyAceTag(&quot;phdBall&quot;, &quot;consed&quot;,</span>
<span class="fc" id="L207">	                    DateUtil.getCurrentDate(), &quot;../phdball_dir/&quot;+phdFile.getName()));</span>
<span class="fc" id="L208">				aceWriter.close();</span>
			}finally{
				//delete our tmpdir
				//no matter what
<span class="pc" id="L212">				IOUtil.recursiveDelete(tmpDir);</span>
<span class="fc" id="L213">			}</span>
<span class="nc" id="L214">		} catch (IOException e) {</span>
			// TODO Auto-generated catch block
<span class="nc" id="L216">			e.printStackTrace();</span>
<span class="fc" id="L217">		}</span>
		
<span class="fc" id="L219">	}</span>

	protected void visitBeginReference(String key) {
		// TODO Auto-generated method stub
		
<span class="fc" id="L224">	}</span>

	protected void visitEndReference() {
		// TODO Auto-generated method stub
		
<span class="fc" id="L229">	}</span>
	

	@Override
    protected StreamingIterator&lt;? extends Trace&gt; createIteratorFor(
            FastqFileDataStore datastore) throws DataStoreException {
<span class="nc" id="L235">	    return new FastqConsedPhdAdaptedIterator( </span>
<span class="nc" id="L236">                    datastore.iterator(),</span>
<span class="nc" id="L237">                        datastore.getFile().get(), </span>
                        phdDate);
    }

  
	@Override
	protected StreamingIterator&lt;PhdReadRecord&gt; createSffIterator(File sffFile)
			throws DataStoreException, IOException {
<span class="nc" id="L245">		return new FlowgramConsedPhdAdaptedIterator(</span>
<span class="nc" id="L246">				 new SffFileDataStoreBuilder(sffFile).hint(DataStoreProviderHint.ITERATION_ONLY).build().iterator(),</span>
                sffFile,
                phdDate);
	}

	@Override
	protected StreamingIterator&lt;PhdReadRecord&gt; createFastaIterator(
			File file) throws DataStoreException {
<span class="pc bpc" id="L254" title="1 of 2 branches missed.">		if(chromatDir ==null){</span>
			try {
<span class="fc" id="L256">				NucleotideFastaDataStore datastore = new NucleotideFastaFileDataStoreBuilder(file)</span>
<span class="fc" id="L257">																.hint(DataStoreProviderHint.ITERATION_ONLY)</span>
<span class="fc" id="L258">																.build();</span>
<span class="fc" id="L259">				return new QualFastaConsedPhdAdaptedIterator(</span>
<span class="fc" id="L260">		                datastore.iterator(),</span>
		                file,
<span class="fc" id="L262">		                phdDate, PhredQuality.valueOf(30));</span>
<span class="nc" id="L263">	        } catch (IOException e) {</span>
<span class="nc" id="L264">				throw new DataStoreException(&quot;error reading fasta file &quot;+ file.getAbsolutePath(),e);</span>
			}
		}else{
			try {
	        	//there should only be a few sanger traces so we can take the memory 
	        	//hit and store it all in memory
<span class="nc" id="L270">				NucleotideFastaDataStore datastore = new NucleotideFastaFileDataStoreBuilder(file)</span>
<span class="nc" id="L271">															.build();</span>
<span class="nc" id="L272">				return new ChromatDirFastaConsedPhdAdaptedIterator(</span>
<span class="nc" id="L273">						datastore.iterator(),</span>
						file,
<span class="nc" id="L275">		                phdDate, PhredQuality.valueOf(30),</span>
		                chromatDir);
<span class="nc" id="L277">	        } catch (IOException e) {</span>
<span class="nc" id="L278">				throw new DataStoreException(&quot;error reading fasta file for chromatogram &quot;+ file.getAbsolutePath(),e);</span>
			}
		}
	}

	
	
}
</pre><div class="footer"><span class="right">Created with <a href="http://www.eclemma.org/jacoco">JaCoCo</a> 0.7.5.201505241946</span></div></body></html>